<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Arman Shahrisa" />

<meta name="date" content="2017-08-19" />

<title>cbaf</title>






<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">cbaf</h1>
<h4 class="author"><em>Arman Shahrisa</em></h4>
<h4 class="date"><em>2017-08-19</em></h4>



<p>cbaf package facilitates working with high-throughput data stored on <a href="http://www.cbioportal.org/" class="uri">http://www.cbioportal.org/</a>. The official Bioconductor package that is designed for obtaining data from cBioPortal in R, is <code>cgdsr</code>. To obtain data with this package, users have to pass a multistep procedure. Besides, the index of cancers and their subgroups changes frequently, which in turn, requires changing the R code. cbaf makes this procedure automated for <strong>RNA-seq</strong>, <strong>microRNA-seq</strong>, <strong>microarray</strong> and <strong>methylation</strong> data. In addition, comparing the genetic data across multiple cancer studies/cancer study subgroups becomes much faster and easier. The results are stored as excel file(s) and multiple heatmaps.</p>
<p>The package consists of six main functions: <code>availableData()</code>, <code>obtainOneStudy()</code>, <code>obtainMultipleStudies()</code>, <code>automatedStatistics()</code>, <code>heatmapOutput()</code> and <code>xlsxOutput()</code>.</p>
<p>There are two other functions, <code>processOneStudy()</code> and <code>processMultipleStudies()</code>, that combine multiple functions in order to automatically execute multiple functions following one another. Please use these two functions instead of exceting main functions one by one. This allows functions to work with higher efficiency.</p>
<div id="availabledata" class="section level2">
<h2>availableData()</h2>
<p>This function scans all the cancer studies to examine presence of <em>RNA-seq</em>, <em>microRNA-seq</em>, <em>microarray</em> and <em>methylation</em> data. It requires a name to label the output result.</p>
<pre><code>availableData(&quot;name&quot;)</code></pre>
<p>There is another argument that enables the function to save the result as an excel file. The default value is <code>excelFile = TRUE</code>. To prevent saving excel file, set <code>excelFile = FALSE</code>.</p>
<pre><code>availableData(&quot;name&quot;, excelFile = FALSE)</code></pre>
<p>At first, if <code>excelFile == TRUE</code>, the function starts looking into the working directory for an excel file with the given name. If it finds the file, the following message is printed, asking the user whether or not it should proceeds.</p>
<pre><code>[1] &quot;An excel file with the given name already exists in the working directory. Proceeding will cause the function to overwrite the file! If you wish to run the function again with a different given name, please type no .&quot;
Proceed anyway and overwrite the file? (yes/no):      no</code></pre>
<p>If the answer is no, the function prints a message to inform the user and then stops processing.</p>
<pre><code>[1] &quot;--- Function 'availableData()' was skipped. ---&quot;</code></pre>
<p>If the user types yes or there is no excel file with the given name, the function continues with printing a message to inform the user it is scanning all the available cancer studies. Then, a progress bar will appear in the next line.</p>
<pre><code>[1] &quot;Cheching the available data for every cancer study&quot;
  |                                                                   |   0%
</code></pre>
<p>The progress bar gradually proceeds until it reaches 100%.</p>
<pre><code>[1] &quot;Cheching the available data for every cancer study&quot;
  |===================================================================| 100%</code></pre>
<p>Upon finishing, the output is accessible with the given name as a variable. if <code>excelFile = TRUE</code>, an excel file containing the results is also generated in the present directory. Both variable and excel file contain different columns: cancer_study_id, cancer_study_name, RNA.seq, microRNA.seq, microarray of mRNA, microarray of miRNA, methylation and description.</p>
</div>
<div id="obtainonestudy" class="section level2">
<h2>obtainOneStudy()</h2>
<p>This function obtains and stores the supported data for at least one group of genes across multiple subgroups of a cancer study. It can check whether or not all genes are included in different subgroups of a cancer study and, if not, looks for the alternative gene names.</p>
<p>It requires at least four arguments:</p>
<ul>
<li><p><strong>genesList</strong>, a list that contains at least one gene group. There is no limit for the number of gene groups, users can set as many as gene groups they desire.</p></li>
<li><p><strong>submissionName</strong>, a character string containing name of interest. It is used for naming the process.</p></li>
<li><p><strong>studyName</strong>, a character string showing the desired cancer name. It is an standard cancer study name that can be found on cbioportal.org, such as <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>.</p></li>
<li><p><strong>desiredTechnique</strong>, one of the five supported high-throughput studies: <code>RNA-seq</code>, <code>microRNA-Seq</code>, <code>microarray.mRNA</code>, <code>microarray.microRNA</code> or <code>methylation</code>.</p></li>
<li><p><strong>desiredCaseList</strong> a numeric vector that contains the index of desired cancer subgroups, assuming the user knows index of desired subgroups. If not, desiredCaseList must be set as ‘none’, function will show the available subgroups and ask the user to enter the desired ones during the process. The default value is ‘none’.</p></li>
<li><p><strong>validateGenes</strong> a logical value that, if set to be ‘TRUE’, function will check each cancer subgroup to find whether or not every gene has a record. If the subgroup doesn’t have a record for the specific gene, function checks for alternative gene names that cbioportal might use instead of the given gene name.</p></li>
</ul>
<p>In the following example, <em>genes</em> consists of two gene groups K.demethylases and K.acetyltransferases, <em>submissionName</em> is <code>test</code>, <em>cancername</em> is <code>Breast Invasive Carcinoma (TCGA, Cell 2015)</code> and the <em>desiredTechnique</em> is <code>RNA-seq</code>.</p>
<pre><code>genes &lt;- list(K.demethylases = c(&quot;KDM1A&quot;, &quot;KDM1B&quot;, &quot;KDM2A&quot;), K.acetyltransferases = c(&quot;CLOCK&quot;, &quot;CREBBP&quot;, &quot;ELP3&quot;, &quot;EP300&quot;))

obtainOneStudy(genes, &quot;test&quot;, &quot;Breast Invasive Carcinoma (TCGA, Cell 2015)&quot;, &quot;RNA-seq&quot;)
</code></pre>
<p>Here is another example that shows how the index of desired subgroups can be used:</p>
<pre><code>obtainOneStudy(genes, &quot;test&quot;, &quot;Breast Invasive Carcinoma (TCGA, Cell 2015)&quot;, &quot;RNA-seq&quot;, desiredCaseList = c(2,3,4,5))</code></pre>
<p>First, obtainOneStudy() checks whether the requested data has already been obtained. If it can find it, function prints the following message and then stops further processing.</p>
<pre><code>[1] &quot;--- Function 'obtainOneStudy()' was skipped: the requested data already exist ---&quot;</code></pre>
<p>If there is a change in function input data, e.g. gene names, or the requested data has not been obtained, assuming <code>desired.case.list = &quot;none&quot;</code>, all subgroups of the requested cancer study appear on console, asking the user to choose the index of desired subgroups:</p>
<pre><code>[1] &quot;Please enter the numeric index of desired case list(s) for Breast Invasive Carcinoma (TCGA, Cell 2015) separated by comma. For instance two indexes 1 and 2 should be enterd as 1, 2&quot;
 [1] &quot;1. All Complete Tumors&quot;                                &quot;2. All Tumors&quot;                                        
 [3] &quot;3. ER+ breast tumors&quot;                                  &quot;4. ER- breast tumors&quot;                                 
 [5] &quot;5. Her2-positive breast tumors&quot;                        &quot;6. Invasive Ductal Cancer (Luminal A)&quot;                
 [7] &quot;7. Invasive Ductal Cancer (PAM50 Basal-like)&quot;          &quot;8. Invasive Ductal Cancer (PAM50 Her2-enriched)&quot;      
 [9] &quot;9. Invasive Ductal Cancer (PAM50 Luminal B)&quot;           &quot;10. Invasive Lobular Cancer&quot;                          
[11] &quot;11. Invasive Lobular Cancer (Luminal A)&quot;               &quot;12. Mixed IDC/ILC breast cancer&quot;                      
[13] &quot;13. Other histologies breast cancer&quot;                   &quot;14. Sequenced Tumors&quot;                                 
[15] &quot;15. TCGA Freeze 2015&quot;                                  &quot;16. Triple-negative breast tumors&quot;                    
[17] &quot;17. Tumor Samples with CNA data&quot;                       &quot;18. Tumor Samples with methylation data (HM27)&quot;       
[19] &quot;19. Tumor Samples with methylation data (HM450)&quot;       &quot;20. Tumor Samples with mRNA data (Agilent microarray)&quot;
[21] &quot;21. Tumor Samples with mRNA data (RNA Seq V2)&quot;         &quot;22. Tumor Samples with RPPA data&quot;                     
[23] &quot;23. Tumor Samples with sequencing and CNA data&quot;</code></pre>
<pre><code>Enter the numeric index(es):                  2,3,4,5</code></pre>
<p>Then, it starts to get the data and informs the user with a progress bar.</p>
<pre><code>[1] &quot;*** Obtaining the requested data for test ***&quot;
|===================================================================| 100%</code></pre>
<p>Output is stored as a <em>BiocFileCache</em> object, which its name is combination of <code>bfc_</code> and <em>submissionName</em>. Accordingly, in our example the BiocFileCache name is <code>bfc_test</code>.</p>
</div>
<div id="obtainmultiplestudies" class="section level2">
<h2>obtainMultipleStudies()</h2>
<p>This function obtains and stores the supported data for at least one group of genes across multiple cancer studies. It can check whether or not all genes are included in each cancer study and, if not, it looks for the alternative gene names.</p>
<p>It requires at least four arguments:</p>
<ul>
<li><p><strong>genes</strong>, a list that contains at least one group of genes. There is no limit for the number of gene groups, users can set as many as gene groups they desire.</p></li>
<li><p><strong>submissionName</strong>, a character string containing name of interest. It is used for naming the process.</p></li>
<li><p><strong>cancernames</strong>, a character vector or a matrix possessing names of desired cancer studies. The character vector contains standard cancer names that can be found on cbioportal.org, such as <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>. Alternatively, a matrix can be used if user prefers user-defined cancer names. In this case, the first column of matrix comprises the standard cancer names while the second column must contain the desired cancer names.</p></li>
<li><p><strong>desiredTechnique</strong>, one of the five supported high-throughput studies: <code>RNA-seq</code>, <code>microRNA-Seq</code>, <code>microarray.mRNA</code>, <code>microarray.microRNA</code> or <code>methylation</code>.</p></li>
</ul>
<p>Function also contains two other options:</p>
<ul>
<li><p><strong>cancerCode</strong>, if <code>TRUE</code>, will force the function to use the standard abbreviated cancer names instead of complete cancer names. For example, <code>laml_tcga_pub</code> is the shortened name for <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>.</p></li>
<li><p><strong>validateGenes</strong> , if <code>TRUE</code>, causes the function to check all cancer studies to find which genes from the input data are available. In addition, function checks for alternative gene names that cbioportal might use instead of the given gene name.</p></li>
</ul>
<p>In the following example, <em>genes</em> consists of two gene groups K.demethylases and K.acetyltransferases, <em>submissionName</em> is <code>test</code>, <em>cancername</em> has complete name of five cancer studies and the desired high-throughput study is <code>RNA-seq</code>.</p>
<pre><code>genes &lt;- list(K.demethylases = c(&quot;KDM1A&quot;, &quot;KDM1B&quot;, &quot;KDM2A&quot;), K.acetyltransferases = c(&quot;CLOCK&quot;, &quot;CREBBP&quot;, &quot;ELP3&quot;, &quot;EP300&quot;))


cancernames &lt;- c(&quot;Acute Myeloid Leukemia (TCGA, Provisional)&quot;, &quot;Adrenocortical Carcinoma (TCGA, Provisional)&quot;, &quot;Bladder Urothelial Carcinoma (TCGA, Provisional)&quot;, &quot;Brain Lower Grade Glioma (TCGA, Provisional)&quot;, &quot;Breast Invasive Carcinoma (TCGA, Provisional)&quot;)


obntainMultipleStudies(genes, &quot;test2&quot;, cancernames, &quot;RNA-seq&quot;)</code></pre>
<p>The following code shows how to create a matrix with the desired names for <code>cancernames</code>:</p>
<pre><code>cancernames &lt;- matrix(c(&quot;Acute Myeloid Leukemia (TCGA, Provisional)&quot;, &quot;acute myeloid leukemia&quot;, &quot;Adrenocortical Carcinoma (TCGA, Provisional)&quot;, &quot;adrenocortical carcinoma&quot;, &quot;Bladder Urothelial Carcinoma (TCGA, Provisional)&quot;, &quot;bladder urothelial carcinoma&quot;, &quot;Brain Lower Grade Glioma (TCGA, Provisional)&quot;, &quot;brain lower grade glioma&quot;, &quot;Breast Invasive Carcinoma (TCGA, Provisional)&quot;,  &quot;breast invasive carcinoma&quot;), nrow = 5, ncol=2 , byrow = TRUE)</code></pre>
<p>An example of how more options for <code>processMultipleStudies()</code> can be altered:</p>
<pre><code>ObtainMultipleStudies(genes, &quot;test2&quot;, cancernames, &quot;RNA-seq&quot;, cancerCode = TRUE, validateGenes = FALSE)</code></pre>
<p>Function starts by checking whether or not the requested data has already been obtained and, if the requested data exists, it prints the following message and then prevents further processing.</p>
<pre><code>[1] &quot;--- Function 'obtainMultipleStudies()' was skipped: the requested data already exist ---&quot;</code></pre>
<p>If there is a change in function input data, e.g. gene names, or the requested data has not been obtained, it proceeds by getting the data and informs the user with a progress bar.</p>
<pre><code>[1] &quot;*** Obtaining the requested data for test2 ***&quot;
|===================================================================| 100%</code></pre>
<p>The output is stored as a BiocFileCache object, which its name is combination of <code>bfc_</code> and <em>submissionName</em>. Accordingly, in our example the BiocFileCache name is <code>bfc_test2</code>.</p>
</div>
<div id="automatedstatistics" class="section level2">
<h2>automatedStatistics()</h2>
<p>The function calculates the statistics of the data obtained by <code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code> functions. Based on user’s preference, these statistics can include <em>frequency percentage</em>, <em>frequency ratio</em>, <em>mean value</em> and <em>median value</em> of samples greater than specific value. Furthermore, it can look for the genes that comprise the highest values in each cancer and list the top 5 genes for <em>frequency percentage</em>, <em>mean value</em> and <em>median value</em>.</p>
<p>It requires at least two arguments:</p>
<ul>
<li><p><strong>submissionName</strong>, a character string containing name of interest. It is used for naming the process and should be the same as submissionName for either of <code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code> functions.</p></li>
<li><p><strong>obtainedDataType</strong>, a character string that identifies the type of input data produced by the previous function. Two options are available: <code>single study</code> for <code>obtainOneStudy()</code> and <code>multiple studies</code> for <code>obtainMultipleStudies()</code>. The function uses <code>obtainedDataType</code> and <code>submissionName</code> to construct the name of the BiocFileCach object and then finds the appropriate data inside it. Default value is <code>multiple studies</code>.</p></li>
</ul>
<p>Function also contains four other options:</p>
<ul>
<li><p><strong>calculate</strong>, a character vector that contains the desired statistical procedures. Default input is <code>c(&quot;frequencyPercentage&quot;, &quot;frequencyRatio&quot;, &quot;meanValue&quot;, &quot;medianValue&quot;)</code>. This will tell the function to compute the followings:</p>
<ul>
<li><p><em>frequencyPercentage</em>, which is the percentage of samples having the value greather than specific cutoff divided by the total sample size for every study / study subgroup</p></li>
<li><p><em>frequency ratio</em>, which shows the number of selected samples divided by the total number of samples that give the frequency percentage. It shows the selected and total sample sizes.</p></li>
<li><p><em>Mean Value</em>, which contains mean value of selected samples for each study.</p></li>
<li><p><em>Median Value</em>, which shows the median value of selected samples for every study.</p></li>
</ul></li>
<li><p><strong>topGenes</strong>, a logical value that, if set as TRUE, causes the function to create three data.frame that contain the five top genes for each cancer. To get all the three data.frames, <em>frequencyPercentage</em>, <em>meanValue</em> and <em>median</em> must have been included for <strong>calculate</strong>.</p></li>
<li><p><strong>cutoff</strong>, a number used to limit samples to those that are greater than this number (cutoff). The default value for methylation data is 0.6 while gene expression studies use default value of 2. For methylation studies, it is <em>observed/expected ratio</em>, for the rest, it is <em>z-score</em>. To change the cutoff to any desired number, change the option to <code>cutoff = desiredNumber</code> in which desiredNumber is the number of interest.</p></li>
<li><p><strong>round</strong>, a logical value that forces the function to round all the calculated values to two decimal places. The default value is <code>TRUE</code>.</p></li>
</ul>
<p>In the following example, <em>submissionName</em> is <code>test</code>, and the <em>obtainedDataType</em> is <code>multiple studies</code>.</p>
<pre><code>automatedStatistics(&quot;test&quot;, obtainedDataType = &quot;multiple studies&quot;)
</code></pre>
<p>The following is the same as previous but excludes <em>mean value</em> and <em>median value</em>. Note that top genes for these two statistics will also be skipped.</p>
<pre><code>automatedStatistics(&quot;test&quot;, obtainedDataType = &quot;multiple studies&quot;, calculate = c(&quot;frequencyPercentage&quot;, &quot;frequencyRatio&quot;))
</code></pre>
<p>Function starts by checking whether or not output data for the previous function exists. If not, an error message will appear:</p>
<pre><code>Error: Please run one of the obtainSingleStudy() or obtainMultipleStudies() functions first</code></pre>
<p>If it can find the desired data but one of the two previous function (either <code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code>) was skipped before, function then looks for its own output. If the requested output already exists, function will stop further processing, printing the following message:</p>
<pre><code>[1] &quot;--- Function 'automatedStatistics()' was skipped: the requested data already exist ---&quot;</code></pre>
<p>If there is a change in function input data, e.g. change in cutoff value, or the output result of automatedStatistics() does not exist, function continues by computing the requested statistics and informs the user with a progress bar.</p>
<pre><code>[1] &quot;*** Performing the requested statistical analyses for test2 ***&quot;
|===================================================================| 100%</code></pre>
<p>The output is stored as a new section of the previous BiocFileCache object which is created by one of the two former functions (<code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code>).</p>
</div>
<div id="heatmapoutput" class="section level2">
<h2>heatmapOutput()</h2>
<p>This function prepares heatmap for <em>frequency percentage</em>, <em>mean value</em> and <em>median value</em> data provided by <code>automatedStatistics()</code> function. Heatmaps for every gene group are stored in separate folder.</p>
<p>It requires at least one arguments:</p>
<ul>
<li><strong>submissionName</strong>, a character string containing name of interest. It is used for naming the process and should be the same as submissionName for either of <code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code> functions.</li>
</ul>
<p>Function also contains thirteen other options:</p>
<ul>
<li><p><strong>shortenStudyNames</strong> a logical value that causes the function to remove the last part of cancer names aiming to shorten them. The removed segment usually contains the name of scientific group that has conducted the experiment.</p></li>
<li><p><strong>genelimit</strong> if large number of genes exist in at least one gene group, this option can be used to limit the number of genes that are shown on heatmap. For instance, <code>genelimit=50</code> will limit the heatmap to 50 genes that show the most variation across multiple study / study subgroups. The default value is <code>none</code>.</p></li>
<li><p><strong>resolution</strong> This option can be used to adjust the resolution of the output heatmaps as ‘dot per inch’. The defalut resolution is 600.</p></li>
<li><p><strong>RowCex</strong> a number that specifies letter size in heatmap row names.</p></li>
<li><p><strong>ColCex</strong> a number that specifies letter size in heatmap column names.</p></li>
<li><p><strong>heatmapMargines</strong> a numeric vector that is used to set heatmap margins. The default value is <code>heatmapMargines=c(15,07)</code>.</p></li>
<li><p><strong>angleForYaxisNames</strong> a number that determines the angle with which the studies/study subgroups names are shown on heatmaps. The default value is 45 degree.</p></li>
<li><p><strong>heatmapColor</strong> a character string that defines heatmap color. The default value is “RdBu”. “redgreen” is also a popular color in genomic studies. To see the rest of colors, please type <code>library(RColorBrewer)</code> and then <code>display.brewer.all()</code>.</p></li>
<li><p><strong>reverseColor</strong> a logical value that reverses the color gradient for heatmap(s).</p></li>
<li><p><strong>transposedHeatmap</strong> a logical value that transposes heatmap rows to columns and vice versa.</p></li>
<li><p><strong>simplify</strong> a logical value that tells the function whether or not to change values under <em>simplifictionCuttoff</em> to zero. The purpose behind this option is to facilitate seeing the candidate genes. Therefore, it is not suited for publications.</p></li>
<li><p><strong>simplifictionCuttoff</strong> a logical value that, if <code>simplify.visulization = TRUE</code>, needs to be set as a desired cutoff for <em>simplify.visulization</em>. It has the same unit as <em>cutoff</em>.</p></li>
<li><p><strong>genesToDrop</strong> a character vector. Gene names within this vector will be omitted from heatmap.</p></li>
</ul>
<p>In the following example, <em>submissionName</em> is <code>test</code>.</p>
<pre><code>heatmapOutput(submissionName = &quot;test&quot;)
</code></pre>
<p>The following is the same as previous but uses more options.</p>
<pre><code>heatmapOutput(&quot;test&quot;, shortenStudyNames = TRUE, heatmapMargines = c(13,5), heatmapColor = &quot;redgreen&quot;, genesToDrop = c(&quot;PVT1&quot;, &quot;SNHG6&quot;), RowCex = 1, ColCex = 1, reverseColor = FALSE)
</code></pre>
<p>Function starts by checking whether or not output data of the previous function exists. If not, an error message will appear:</p>
<pre><code>Error: Please run one of the obtainSingleStudy() or obtainMultipleStudies() functions and then the automatedStatistics() function first.</code></pre>
<p>Otherwise, it continues by preparing and storing the requested heatmaps and informs the user with a progress bar.</p>
<pre><code>[1] &quot;*** Preparing the requested heatmaps for test ***&quot;
|===================================================================| 100%</code></pre>
<p>If there is no change in the options or the requested heatmaps already exist, it avoids storing the heatmaps. The number of skipped heatmaps is then printed:</p>
<pre><code>&quot;--- 1 out of 1 heatmaps was skipped: It already exists. ---&quot;</code></pre>
</div>
<div id="xlsxoutput" class="section level2">
<h2>xlsxOutput()</h2>
<p>This function exports the output of <code>automatedStatistics()</code> and the <em>gene validation</em> result of one of the <code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code> functions as an excel file. For every gene group, an excel file will be generated and stored in the same folder as heatmaps.</p>
<p>It requires one argument:</p>
<ul>
<li><strong>submissionName</strong>, a character string containing name of interest. It is used for naming the process and should be the same as submissionName for either of <code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code> functions.</li>
</ul>
<p>In the following example, <em>submissionName</em> is <code>test</code>.</p>
<pre><code>xlsxOutput(&quot;test&quot;)
</code></pre>
<p>Function starts by checking whether or not output data of the previous function exists. If not, an error message will appear:</p>
<pre><code>Error: Please run one of the obtainSingleStudy() or obtainMultipleStudies() functions and then the automatedStatistics() function first.</code></pre>
<p>Otherwise, it continues by preparing and storing the requested heatmaps and informs the user with a progress bar.</p>
<pre><code>[1] &quot;*** Preparing the requested excel file(s) for test ***&quot;
|===================================================================| 100%</code></pre>
<p>If there is no change in the input parameters or the requested excel files already exist, it avoids rewriting them. The number of skipped excel files is then printed:</p>
<pre><code>&quot;--- 1 out of 1 excel file was skipped: It already exists. ---&quot;</code></pre>
</div>
<div id="processonestudy" class="section level2">
<h2>processOneStudy()</h2>
<p>This function combines four of the mentioned functions for the ease of use. It is recommended that users only use this parent function to obtain and process gene data across multiple subsections of a cancer study so that child functions work with maximum efficiency. <code>processOneStudy()</code> uses the following functions:</p>
<ul>
<li><strong>obtainOneStudy()</strong></li>
<li><strong>automatedStatistics()</strong></li>
<li><strong>heatmapOutput()</strong></li>
<li><strong>xlsxOutput()</strong></li>
</ul>
<p>The function code along with all available options is as follows:</p>
<pre><code>processOneStudy(genesList, submissionName, studyName, desiredTechnique, desiredCaseList = FALSE, validateGenes = TRUE, calculate = c(&quot;frequencyPercentage&quot;, &quot;frequencyRatio&quot;, &quot;meanValue&quot;, &quot;medianValue&quot;), cutoff=NULL, round=TRUE, topGenes = TRUE, shortenStudyNames = TRUE, genelimit = &quot;none&quot;, resolution = 600, RowCex = 0.8, ColCex = 0.8, heatmapMargines = c(10,10), angleForYaxisNames = 45, heatmapColor = &quot;RdBu&quot;, reverseColor = TRUE, transposedHeatmap = FALSE, simplify = FALSE, simplifictionCuttoff = FALSE, genesToDrop = NULL)</code></pre>
<p>To get more information about the function options, please refer to the child function to whom they correspond, for example <code>genesList</code> lies within <code>obtainMultipleStudies()</code> function. The following is an example showing how this function can be used:</p>
<pre><code>genes &lt;- list(K.demethylases = c(&quot;KDM1A&quot;, &quot;KDM1B&quot;, &quot;KDM2A&quot;, &quot;KDM2B&quot;, &quot;KDM3A&quot;, &quot;KDM3B&quot;, &quot;JMJD1C&quot;, &quot;KDM4A&quot;), K.methyltransferases = c(&quot;SUV39H1&quot;, &quot;SUV39H2&quot;, &quot;EHMT1&quot;, &quot;EHMT2&quot;, &quot;SETDB1&quot;, &quot;SETDB2&quot;, &quot;KMT2A&quot;, &quot;KMT2A&quot;))

processOneStudy(genes, &quot;test&quot;, &quot;Breast Invasive Carcinoma (TCGA, Cell 2015)&quot;, &quot;RNA-seq&quot;, desiredCaseList = c(3,4), calculate = c(&quot;frequencyPercentage&quot;,  &quot;frequencyRatio&quot;))
</code></pre>
<p>The output excel files and heatmaps are stored in separate folders for every gene group. Ultimately, all the folders are located inside another folder, which it’s name is the combination of <em>submissionName</em> and “output for single study”, for example in our example it is: “test output for single study”.</p>
</div>
<div id="processmultiplestudies" class="section level2">
<h2>processMultipleStudies()</h2>
<p>This function combines four of the mentioned above functions for the ease of use. It is recommended that users only use this parent function to obtain and process gene data across multiple cancer studies for maximum efficiency. <code>processMultipleStudies()</code> uses the following functions:</p>
<ul>
<li><strong>obtainMultipleStudies()</strong></li>
<li><strong>automatedStatistics()</strong></li>
<li><strong>heatmapOutput()</strong></li>
<li><strong>xlsxOutput()</strong></li>
</ul>
<p>The function code along with all available options is as follows:</p>
<pre><code>processMultipleStudies(genesList, submissionName, studiesNames, desiredTechnique, cancerCode = FALSE, validateGenes = TRUE, calculate = c(&quot;frequencyPercentage&quot;, &quot;frequencyRatio&quot;, &quot;meanValue&quot;, &quot;medianValue&quot;), cutoff=NULL, round=TRUE, topGenes = TRUE, shortenStudyNames = TRUE, genelimit = &quot;none&quot;, resolution = 600, RowCex = 0.8, ColCex = 0.8, heatmapMargines = c(10,10), angleForYaxisNames = 45, heatmapColor = &quot;RdBu&quot;, reverseColor = TRUE, transposedHeatmap = FALSE, simplify = FALSE, simplifictionCuttoff = FALSE, genesToDrop = NULL)</code></pre>
<p>To obtain more information about the function options, please refer to the child function to whom they correspond, for example <code>genesList</code> lies within <code>ontainOneStudy()</code> function. The following example shows how this function can be used:</p>
<pre><code>genes &lt;- list(K.demethylases = c(&quot;KDM1A&quot;, &quot;KDM1B&quot;, &quot;KDM2A&quot;, &quot;KDM2B&quot;, &quot;KDM3A&quot;, &quot;KDM3B&quot;, &quot;JMJD1C&quot;, &quot;KDM4A&quot;), K.methyltransferases = c(&quot;SUV39H1&quot;, &quot;SUV39H2&quot;, &quot;EHMT1&quot;, &quot;EHMT2&quot;, &quot;SETDB1&quot;, &quot;SETDB2&quot;, &quot;KMT2A&quot;, &quot;KMT2A&quot;))

studies &lt;- c(&quot;Acute Myeloid Leukemia (TCGA, Provisional)&quot;, &quot;Adrenocortical Carcinoma (TCGA, Provisional)&quot;)

processMultipleStudies(genes, &quot;test2&quot;, studies, &quot;RNA-seq&quot;, calculate = c(&quot;frequencyPercentage&quot;, &quot;frequencyRatio&quot;), heatmapMargines = c(15,10))
</code></pre>
<p>The output excel file and heatmaps are stored in separate folders for every gene group. Ultimately, all the folders are located inside another folder, which it’s name that is the combination of <em>submissionName</em> and “output for multiple studies”, for example “test output for multiple studies”.</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
