<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="date" content="2017-05-31" />

<title>cBioAutomatedTools</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">cBioAutomatedTools</h1>
<h4 class="date"><em>2017-05-31</em></h4>



<p>cBioAutomatedTools package facilitates working with high-throughput data stored on <a href="http://www.cbioportal.org/" class="uri">http://www.cbioportal.org/</a>. The official Bioconductor package that is designed for obtaining data from cBioPorta in R, is <code>cgdsr</code>. To obtain data with this package, users have to pass a multi-step procidure. Besides, the index of cancers and thier subgroups might change frequently, which in turn, requires changing the R code. <code>cBioAutomatedTools</code> makes this procedure automated for <strong>RNA-seq</strong>, <strong>microRNA-seq</strong>, <strong>microarray</strong> and <strong>methylation</strong> data types. In addition, comparing the genetic data from different cancer studies/cancer study subgroups becomes much faster and easier. Further, the results are presented as an excel file and multiple heatmaps.</p>
<p>The package consists of three functions: <code>available.data.types()</code>, <code>process.data.for.cancers()</code> and <code>process.data.for.a.cancer()</code>.</p>
<div id="available.data.types" class="section level2">
<h2>available.data.types()</h2>
<p>This function ,requires no input, scans all the cancer studies to find which of the <em>RNA-seq</em>, <em>microRNA-seq</em>, <em>microarray</em> and <em>methylation</em> data types exist for each study. To use this function, simply type:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="kw">available.data.types</span>()</code></pre></div>
<p>The code starts with printing a message in the R console to inform the user it is scanning all the available cancer studies. In the next line, a progress bar will appear.</p>
<pre><code>[1] &quot;Cheching which data types are available for each cancer&quot;
  |                                                                 |   0%
</code></pre>
<p>The progress bar gradually proceeds untill it reaches 100%. Until now, 151 cancer studies are registered in cBioPortal. Therefre, the function would requires some time to finish. Upon the finishing, an additional line will be printed in the console that showes the directory in which <em>Available Data Types.xlsx</em> has been saved and the name with which the output is stored in R’s global enviroment (user’s workspace).</p>
<pre><code>[1] &quot;Cheching which data types are available for each cancer&quot;
  |===================================================================| 100%
[1] &quot;An .xlsx file entitled 'Available Data Types Output.xlsx' ,which shows the availabile data types for each cancer study, was saved in Z:/cBioAutomatedTools/Source code/cBioAutomatedTools . In adition, the output is available in the global enviroment as 'available.data.types.output'&quot;</code></pre>
<p>The following table shows how the <em>Available Data Types.xlsx</em> file looks like. Since the complete output is too large, only the first five rows are shown.</p>
<table>
<colgroup>
<col width="2%"></col>
<col width="47%"></col>
<col width="6%"></col>
<col width="8%"></col>
<col width="12%"></col>
<col width="13%"></col>
<col width="8%"></col>
</colgroup>
<thead>
<tr class="header">
<th align="center"></th>
<th align="left"></th>
<th align="center">RNA-seq</th>
<th align="center">MicroRNA-seq</th>
<th align="center">Microarray (mRNA)</th>
<th align="center">Microarray (miRNA)</th>
<th align="center">Methylation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">1</td>
<td align="left">Acinar Cell Carcinoma of the Pancreas (Johns Hopkins, J Pathol 2014)</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
</tr>
<tr class="even">
<td align="center">2</td>
<td align="left">Acute Myeloid Leukemia (TCGA, NEJM 2013)</td>
<td align="center">Available</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">Available</td>
</tr>
<tr class="odd">
<td align="center">3</td>
<td align="left">Acute Myeloid Leukemia (TCGA, Provisional)</td>
<td align="center">Available</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">Available</td>
</tr>
<tr class="even">
<td align="center">4</td>
<td align="left">Adenoid Cystic Carcinoma (FMI, Am J Surg Pathl. 2014)</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
</tr>
<tr class="odd">
<td align="center">5</td>
<td align="left">Adenoid Cystic Carcinoma (MDA, Clin Cancer Res 2015)</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
<td align="center">-</td>
</tr>
</tbody>
</table>
</div>
<div id="process.multiple.studies" class="section level2">
<h2>process.multiple.studies()</h2>
<p>This function Obtains the requested data for the given genes across the multiple cancer studies. It can check whether or not all genes are included in cancer studies and, if not, looks for the alternative gene names. The main part of function calculates frequency percentage, frequency ratio, mean expression and median of samples greather than specific value. Furthermore, it looks for the genes that comprise the highest values in each cancer. The function has been divided into two main segments. The first segment obtaines and processes the desired data while the second segment generates heatmap and excel files.</p>
<p>The function requires at least three variables:</p>
<ul>
<li><strong>genes</strong>, a list that contains at least one group of genes</li>
<li><strong>cancernames</strong>, a character vector or a matrix posessing names of desired cancer studies. The character vector containes standard cancer names that can be found on cbioportal.org, such as <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>. Alternatively, a matrix can be used if users prefer user-defined cancer names, in which the first column of matrix comprises the standard cancer names while the second column must contain the desired cancer names.</li>
<li><strong>high.throughput.data.type</strong>, one of the five supported high-troughput studies: RNA-seq, MicroRNA-seq, microarray.mRNA, microarray.microRNA and methylation</li>
</ul>
<p>In the following example, <em>genes</em> conisists of two gene groups K.demethylases and K.acetyltransferases, <em>cancernames</em> has complete names of five cancer studies and the desired high-throuput study is RNA-seq.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dt">K.demethylases =</span> <span class="kw">c</span>(<span class="st">&quot;KDM1A&quot;</span>, <span class="st">&quot;KDM1B&quot;</span>, <span class="st">&quot;KDM2A&quot;</span>), <span class="dt">K.acetyltransferases =</span> <span class="kw">c</span>(<span class="st">&quot;CLOCK&quot;</span>, <span class="st">&quot;CREBBP&quot;</span>, <span class="st">&quot;ELP3&quot;</span>, <span class="st">&quot;EP300&quot;</span>))

cancernames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Acute Myeloid Leukemia (TCGA, Provisional)&quot;</span>, <span class="st">&quot;Adrenocortical Carcinoma (TCGA, Provisional)&quot;</span>, <span class="st">&quot;Bladder Urothelial Carcinoma (TCGA, Provisional)&quot;</span>, <span class="st">&quot;Brain Lower Grade Glioma (TCGA, Provisional)&quot;</span>, <span class="st">&quot;Breast Invasive Carcinoma (TCGA, Provisional)&quot;</span>)

<span class="kw">process.multiple.studies</span>(genes, cancernames, <span class="st">&quot;RNA-seq&quot;</span>)</code></pre></div>
<p>The following code shows how to create a matrix with desired names for <code>cancernames</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">cancernames &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="st">&quot;Acute Myeloid Leukemia (TCGA, Provisional)&quot;</span>, <span class="st">&quot;acute myeloid leukemia&quot;</span>, <span class="st">&quot;Adrenocortical Carcinoma (TCGA, Provisional)&quot;</span>, <span class="st">&quot;adrenocortical carcinoma&quot;</span>, <span class="st">&quot;Bladder Urothelial Carcinoma (TCGA, Provisional)&quot;</span>, <span class="st">&quot;bladder urothelial carcinoma&quot;</span>, <span class="st">&quot;Brain Lower Grade Glioma (TCGA, Provisional)&quot;</span>, <span class="st">&quot;brain lower grade glioma&quot;</span>, <span class="st">&quot;Breast Invasive Carcinoma (TCGA, Provisional)&quot;</span>,  <span class="st">&quot;breast invasive carcinoma&quot;</span>), <span class="dt">nrow =</span> <span class="dv">5</span>, <span class="dt">ncol=</span><span class="dv">2</span> , <span class="dt">byrow =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Function containes other options that can be used for a veriety of needs:</p>
<ul>
<li><strong>data.presented.as</strong> tells the function what to compute. The default is <code>c(&quot;Frequency.Percentage&quot;, &quot;Frequency.Ratio&quot;, &quot;Mean.Value&quot;, &quot;Median&quot;)</code> which causes the function to consider all the four possible statistical procedures. Alternatively, users can decrease the number of requested statistical procedures as long as at least one procedure is given to the function.</li>
<li><strong>shorteded.cancer.names</strong> is used to remove the end part of cancer names aiming to shorten them. The removed segment usually contains the name of scientific group that has conducted the experiment.</li>
<li><strong>genelimit</strong> can be employed if large number of genes exists in at least one gene group. This option limits the number of genes to be shown on hitmap. For instance, <code>genelimit=50</code> will limit the heatmap to show 50 genes showing the most variation. The default value is <code>FALSE</code>.</li>
<li><strong>resolution</strong> adjusts the resolution of the output heatmaps as <em>dot per inch</em>. The defalut value is 600.</li>
<li><strong>RowCex</strong> specifies letter size in heatmap row names.</li>
<li><strong>ColCex</strong> specifies letter size in heatmap column names.</li>
<li><strong>heatmapMargines</strong> sets heatmap margins. The default value is <code>heatmapMargines=c(15,07)</code>.</li>
<li><strong>cutoff</strong> used to limit samples to those that are greather than specific number (cutoff). The default value for methylation data is 0.6 while gene expression studies use default value of 2. For methylation studies, the unit is <code>observed/expected ratio</code>, for the rest, it is <code>z-score</code>- function can understand the correct unit. TO change the cutoff to any desired number, change the option to <code>cutoff = desiredNumber</code> in which <em>desiredNumber</em> is the number of interest.</li>
<li><strong>angle.for.heatmap.cancernames</strong> determines the angle with which the cancer names are shown in heatmaps. The default value is 45 degree.</li>
<li><strong>heatmap.color</strong> can be used to set hratmap colors. The default value is “RdBu”. “redgreen” is also a popular color in genomic studies. To see the rest of colors, please type <code>display.brewer.all()</code>.</li>
<li><strong>reverse.heatmap.color</strong> can reverse the color gradiant for heatmap.</li>
<li><strong>rewrite.output.list</strong> This option can be used to modify heatmap, e.g. change margin, without obtaining data from internet again. If set to <code>false</code>, function will use the values previously stored in the global environment (user’s workspace) to draw heatmaps and save an excel file. The default value is <code>TRUE</code>.</li>
<li><strong>round</strong> forces the function to round all the calculated values to two decimal places. The default value is <code>TRUE</code>.</li>
<li><strong>top.genes</strong>, if set as <code>TRUE</code>, cause the function to create three data.frame that contain the gene names with the highest values for each cancer. To get all the three data.frames, <code>Frequency.Percentage</code>, <code>Mean.Value</code> and <code>Median</code> must have been included for <code>data.presented.as</code>.</li>
<li><strong>validate.genes</strong>, if set to be <code>TRUE</code>, function will checks each cancer study to finds whether or not each gene has a record. If a cancer study doesn’t have a record for specific gene, it checks for alternative gene names that cbioportal might use instead of the given gene name.</li>
<li><strong>Use.CancerCode.as.Name</strong>, if set as <code>TRUE</code>, function will use the sandard abbreviated cancer names instead of complete cancer names. For example, <code>laml_tcga_pub</code> is the shortened name for <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>.</li>
<li><strong>simplify.visulization</strong>, if <code>TRUE</code>, function alters the values under <code>simplifiction.cuttoff</code> to zero. It only affects heatmaps to assist finding the candidate genes faster. Therefore, it is not suited for publications.</li>
<li><strong>simplifiction.cuttoff</strong> a logical value that, if <code>simplify.visulization = TRUE</code>, needs to be set as a desired cuttoff for <code>simplify.visulization</code>. It has the same unit as <code>cutoff</code>.</li>
</ul>
<p>An example of how more options for <code>process.multiple.studies()</code> can be altered:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">process.multiple.studies</span>(genes, cancernames, <span class="st">&quot;RNA-seq&quot;</span>, <span class="dt">data.presented.as =</span> <span class="kw">c</span>(<span class="st">&quot;Frequency.Percentage&quot;</span>, <span class="st">&quot;Frequency.Ratio&quot;</span>, <span class="st">&quot;Mean.Value&quot;</span>), <span class="dt">shorteded.cancer.names =</span> <span class="ot">FALSE</span>, <span class="dt">resolution=</span><span class="dv">300</span>, <span class="dt">RowCex=</span><span class="dv">1</span>, <span class="dt">ColCex=</span><span class="dv">1</span>, <span class="dt">heatmapMargines=</span><span class="kw">c</span>(<span class="dv">15</span>,<span class="dv">5</span>), <span class="dt">cutoff=</span><span class="fl">1.5</span>, <span class="dt">angle.for.heatmap.cancernames=</span><span class="dv">30</span>, <span class="dt">heatmap.color =</span> <span class="st">&quot;redgreen&quot;</span>)</code></pre></div>
<div id="first-segment-of-function-obtaining-and-processing-the-requested-data" class="section level3">
<h3>First segment of function: obtaining and processing the requested data</h3>
<p>Just like the previous function, <code>process.multiple.studies()</code> starts by printing that it is obtaining and processing the the requested data. Then, a progress bar will appear in the next line.</p>
<pre><code>[1] &quot;[[[ Obtaining and processing the requested data for K.demethylases genes ]]]&quot;
  |                                                                 |   0%          </code></pre>
<p>The progress bar gradually proceeds to 100%. Large number of genes or cancer studies may prolong the processing time. Upon finishing, another message is printed on console, the function creates and stores a list in the R’s global enviroment with an informative name.</p>
<pre><code>[1] &quot;[[[ Obtaining and processing the requested data for K.demethylases genes ]]]&quot;
  |=================================================================| 100%
[1] &quot;Preparing a 'list' to fill with the requested data&quot;</code></pre>
<p>Based on user’s preference, this list can include some or all of the following groups:</p>
<ul>
<li><strong>Validation.Results</strong>, a character matrix that shows wheather or not different cancer studies possess record for various genes. If a cancer study doesn’t have a record for a gene but uses an alternative name, it is shown in the matrix.</li>
<li><strong>Frequency.Percentage</strong>, a numerical matrix that is obtained by dividing the number of samples with values greather than specific cutoff by the total sample size for each cancer study.</li>
<li><strong>Top.Genes.of.Frequency.Percentage</strong>, a data.frame that showes genes with highest value in Frequency.Percentage. It contains name of the gene showing the highest value for every cancer study plus the corresponding value.</li>
<li><strong>Frequency.Ratio</strong>, a character matrix of fractions. Each fraction contains the number of samples with values greater than specific cutoff as a numerator and the total sample size for each cancer study as a denominator.</li>
<li><strong>Mean.Value</strong>, a numerical matrix that shows mean value of samples greater than specific cutoff.</li>
<li><strong>Top.Genes.of.Genes.Expression</strong>, a data.frame that showes genes with highest value in Mean.Value. It contains name of the gene showing the highest value for every cancer study plus the corresponding value.</li>
<li><strong>Median</strong>, a numerical matrix that shows median value of samples greater than specific cutoff.</li>
<li><strong>Top.Genes.of.Median</strong>, a data.frame that showes genes with highest value in Median. It contains name of the gene showing the highest value for every cancer study plus the corresponding value.</li>
</ul>
</div>
<div id="second-segment-of-function-generating-three-heatmaps-and-one-excel-file" class="section level3">
<h3>Second segment of function: generating three heatmaps and one excel file</h3>
<p>Thse second part of function saves all subgroups of the list, generated by first part of the function, in the different tabs of an excel file. In addition, three heatmaps are generated for Frequency.Percentage, Mean.Value and Median subgroups of the list; based on user’s preference. During the process, a progress bar informs the user about the amount of work that have performed. When finished, a message appears on console that shows the directory where output files can be found.</p>
<pre><code>[1] &quot;[[[ Obtaining and processing the requested data for K.demethylases genes ]]]&quot;
  |=================================================================| 100%
[1] &quot;Preparing a 'list' to fill with the requested data&quot;
[1] &quot;Saving the output files&quot;
  |=================================================================| 100%
[1] &quot;Requested excel and PNG files were saved in Z:/cBioAutomatedTools/Source code/cBioAutomatedTools/1. K-demethylases&quot;</code></pre>
<p>If users intend to modify the characteristics of generated heatmaps, e.g. change color, function can use the information that is previously stored in the global enviroment, generating the outputs files without using the internet. To do so, set the <code>rewrite.output.list</code> to FLASE. For instancce, the following code can be used to change heatmap color to red-green color gradient:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">process.multiple.studies</span>(genes, cancernames, <span class="st">&quot;RNA-seq&quot;</span>, <span class="dt">heatmap.color =</span> <span class="st">&quot;redgreen&quot;</span>, <span class="dt">rewrite.output.list =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
</div>
<div id="process.one.study" class="section level2">
<h2>process.one.study()</h2>
<p>This function Obtains the requested data for the given genes across the multiple subgroups of a cancer study. It is very similar in what it does, to <code>process.multiple.studies()</code> as well as using a two-step procedure.</p>
<p>The function requires at least three variables:</p>
<ul>
<li><strong>genes</strong>, a list that contains at least one group of genes</li>
<li><strong>cancername</strong>, a character string showing the desired cancer name. It is an standard cancer name that can be found on cbioportal.org, such as <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>.</li>
<li><strong>high.throughput.data.type</strong>, one of the five supported high-troughput studies: RNA-seq, MicroRNA-seq, microarray.mRNA, microarray.microRNA and methylation</li>
</ul>
<p>In the following example, <em>genes</em> conisists of two gene groups K.demethylases and K.acetyltransferases, <em>cancername</em> has complete names of five cancer studies and the desired high-throuput study is RNA-seq.</p>
<p>The difference lies in asking the user to chose the desired subroups while the function is working. alternatively, if user knows index of desired subgroups,</p>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
