<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Arman Shahrisa" />

<meta name="date" content="2017-08-16" />

<title>cbaf</title>






<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">cbaf</h1>
<h4 class="author"><em>Arman Shahrisa</em></h4>
<h4 class="date"><em>2017-08-16</em></h4>



<p>cbat package facilitates working with high-throughput data stored on <a href="http://www.cbioportal.org/" class="uri">http://www.cbioportal.org/</a>. The official Bioconductor package that is designed for obtaining data from cBioPortal in R, is <code>cgdsr</code>. To obtain data with this package, users have to pass a multi-step procedure. Besides, the index of cancers and their subgroups might change frequently, which in turn, requires changing the R code. <code>cbaf</code> makes this procedure automated for <strong>RNA-seq</strong>, <strong>microRNA-seq</strong>, <strong>microarray</strong> and <strong>methylation</strong> data. In addition, comparing the genetic data across the different cancer studies/cancer study subgroups becomes much faster and easier. The results are presented as excel file(s) and multiple heatmaps.</p>
<p>The package consists of six main functions: <code>availableData()</code>, <code>obtainOneStudy()</code>, <code>obtainMultipleStudies()</code>, <code>automatedStatistics()</code>, <code>heatmapOutput()</code> and <code>xlsxOutput()</code>.</p>
<p>There are two other functions that combine multiple functions in order to perform automated procedures of obtaining the required data, calculating statistics and exporting the data as heatmaps and excel files: <code>processOneStudy()</code> and <code>processMultipleStudies()</code>.</p>
<div id="availabledata" class="section level2">
<h2>availableData()</h2>
<p>This function scans all the cancer studies to examine presence of <em>RNA-seq</em>, <em>microRNA-seq</em>, <em>microarray</em> and <em>methylation</em> data. It requires a name to lable the output result.</p>
<pre><code>availableData(&quot;name&quot;)</code></pre>
<p>There is another argument that enables the function to save the result as an excel file. The default value is <code>excelFile = TRUE</code>. To prevent saving excel file, set <code>excelFile = FALSE</code>.</p>
<pre><code>availableData(&quot;name&quot;, excelFile = FALSE)</code></pre>
<p>At first, if <code>excelFile == TRUE</code>, the function starts looking for an excel file with the given name in the working directory. If it finds the file, the following message is printed, asking the user whether or not it should proceeds.</p>
<pre><code>[1] &quot;An excel file with the given name already exists in the working directory. Proceeding will cause the function to overwrite the file! If you wish to run the function again with a different given name, please type no .&quot;
Proceed anyway and overwrite the file? (yes/no):      no</code></pre>
<p>If the answer is no, the function prints a message to inform the user and then stops processing.</p>
<pre><code>[1] &quot;--- Function 'availableData()' was skipped. ---&quot;</code></pre>
<p>If the user types yes or there is no excel file with the given name, the function continues with printing a message to inform the user it is scanning all the available cancer studies. Then, a progress bar will appear in the next line.</p>
<pre><code>[1] &quot;Cheching the available data for every cancer study&quot;
  |                                                                   |   0%
</code></pre>
<p>The progress bar gradually proceeds until it reaches 100%.</p>
<pre><code>[1] &quot;Cheching the available data for every cancer study&quot;
  |===================================================================| 100%</code></pre>
<p>Upon finishing, the output is accessible with the given name as a variable. if <code>excelFile = TRUE</code>, an excel file containing the results is also generated in the present directory. Both variable and excel file contain different columns: cancer_study_id, cancer_study_name, RNA.seq, microRNA.seq, microarray of mRNA, microarray of miRNA, methylation and description.</p>
</div>
<div id="obtainonestudy" class="section level2">
<h2>obtainOneStudy()</h2>
<p>This function obtains and stores the supported data for at least one group of genes across multiple subgroups of a cancer study. It can check whether or not all genes are included in cancer studies and, if not, looks for the alternative gene names.</p>
<p>It requires at least four arguments:</p>
<ul>
<li><p><strong>genesList</strong>, a list that contains at least one gene group. There is no limit for the number of gene groups, users can set as many as gene groups they desire.</p></li>
<li><p><strong>submissionName</strong>, a character string containing name of interest. It is used for naming the process.</p></li>
<li><p><strong>studyName</strong>, a character string showing the desired cancer name. It is an standard cancer study name that can be found on cbioportal.org, such as <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>.</p></li>
<li><p><strong>desiredTechnique</strong>, one of the five supported high-throughput studies: RNA-seq, MicroRNA-seq, microarray for mRNA, microarray for microRNA and methylation.</p></li>
</ul>
<p>Function also contains two other options:</p>
<ul>
<li><p><strong>desiredCaseList</strong> a numeric vector that contains the index of desired cancer subgroups, assuming the user knows index of desired subgroups. If not, desiredCaseList must be set as ‘none’, function will show the available subgroups and ask the user to enter the desired ones during the process. The default value is ‘none’.</p></li>
<li><p><strong>validateGenes</strong> a logical value that, if set to be ‘TRUE’, function will check each cancer study to find whether or not each gene has a record. If the given cancer doesn’t have a record for specific gene, it checks for alternative gene names that cbioportal might use instead of the given gene name.</p></li>
</ul>
<p>In the following example, <em>genes</em> consists of two gene groups K.demethylases and K.acetyltransferases, <em>submissionName</em> is <code>test</code>, <em>cancername</em> is <code>Breast Invasive Carcinoma (TCGA, Cell 2015)</code> and the <em>desiredTechnique</em> is <code>RNA-seq</code>.</p>
<pre><code>genes &lt;- list(K.demethylases = c(&quot;KDM1A&quot;, &quot;KDM1B&quot;, &quot;KDM2A&quot;), K.acetyltransferases = c(&quot;CLOCK&quot;, &quot;CREBBP&quot;, &quot;ELP3&quot;, &quot;EP300&quot;))

obtainOneStudy(genes, &quot;test&quot;, &quot;Breast Invasive Carcinoma (TCGA, Cell 2015)&quot;, &quot;RNA-seq&quot;)
</code></pre>
<p>Here is example of how more options can be used with <code>obtainOneStudy()</code>, including index of desired subgroups for <code>desiredCaseList</code>:</p>
<pre><code>process.one.study(genes, cancernames, &quot;RNA-seq&quot;, desired.case.list = c(3,4,5), data.presented.as = c(&quot;Frequency.Percentage&quot;, &quot;Frequency.Ratio&quot;, &quot;Mean.Value&quot;), resolution=300, RowCex=1, ColCex=1, heatmapMargines=c(15,5), cutoff=1.5, angle.for.heatmap.cancernames=30, heatmap.color = &quot;redgreen&quot;)</code></pre>
<p>First, it checks whether the requested data has already been obtained and, if the requested data exists, it prints the following message and then stops further processing.</p>
<pre><code>[1] &quot;--- Function 'obtainOneStudy()' was skipped: the requested data already exist ---&quot;</code></pre>
<p>If there is a change in function input data, e.g. gene names, or the requested data has not been obtained, assuming <code>desired.case.list = &quot;none&quot;</code>, all subgroups of the requested cancer study appear on console, asking the user to chose the index of desired subgroups:</p>
<pre><code>[1] &quot;Please enter the numeric index of desired case list(s) for Breast Invasive Carcinoma (TCGA, Cell 2015) separated by comma. For instance two indexes 1 and 2 should be enterd as 1, 2&quot;
 [1] &quot;1. All Complete Tumors&quot;                                &quot;2. All Tumors&quot;                                        
 [3] &quot;3. ER+ breast tumors&quot;                                  &quot;4. ER- breast tumors&quot;                                 
 [5] &quot;5. Her2-positive breast tumors&quot;                        &quot;6. Invasive Ductal Cancer (Luminal A)&quot;                
 [7] &quot;7. Invasive Ductal Cancer (PAM50 Basal-like)&quot;          &quot;8. Invasive Ductal Cancer (PAM50 Her2-enriched)&quot;      
 [9] &quot;9. Invasive Ductal Cancer (PAM50 Luminal B)&quot;           &quot;10. Invasive Lobular Cancer&quot;                          
[11] &quot;11. Invasive Lobular Cancer (Luminal A)&quot;               &quot;12. Mixed IDC/ILC breast cancer&quot;                      
[13] &quot;13. Other histologies breast cancer&quot;                   &quot;14. Sequenced Tumors&quot;                                 
[15] &quot;15. TCGA Freeze 2015&quot;                                  &quot;16. Triple-negative breast tumors&quot;                    
[17] &quot;17. Tumor Samples with CNA data&quot;                       &quot;18. Tumor Samples with methylation data (HM27)&quot;       
[19] &quot;19. Tumor Samples with methylation data (HM450)&quot;       &quot;20. Tumor Samples with mRNA data (Agilent microarray)&quot;
[21] &quot;21. Tumor Samples with mRNA data (RNA Seq V2)&quot;         &quot;22. Tumor Samples with RPPA data&quot;                     
[23] &quot;23. Tumor Samples with sequencing and CNA data&quot;</code></pre>
<pre><code>Enter the numeric index(es):                  2,3,4,5</code></pre>
<p>Then, it starts to get the data and infroms the user with a progress bar.</p>
<pre><code>[1] &quot;*** Obtaining the requested data for test ***&quot;
|===================================================================| 100%</code></pre>
<p>The output is stored as a BiocFileCache object, which its name is the combination of <code>bfc_</code> and <em>submissionName</em>. Accordingly, in our example the BiocFileCache name is <code>bfc_test</code>.</p>
</div>
<div id="obtainmultiplestudies" class="section level2">
<h2>obtainMultipleStudies()</h2>
<p>This function obtains and stores the supported data for at least one group of genes across multiple cancer studies. It can check whether or not all genes are included in cancer studies and, if not, looks for the alternative gene names.</p>
<p>It requires at least four arguments:</p>
<ul>
<li><p><strong>genes</strong>, a list that contains at least one group of genes. There is no limit for the number of gene groups, users can set as many as gene groups they desire.</p></li>
<li><p><strong>submissionName</strong>, a character string containing name of interest. It is used for naming the process.</p></li>
<li><p><strong>cancernames</strong>, a character vector or a matrix possessing names of desired cancer studies. The character vector contains standard cancer names that can be found on cbioportal.org, such as <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>. Alternatively, a matrix can be used if users prefer user-defined cancer names, in which the first column of matrix comprises the standard cancer names while the second column must contain the desired cancer names.</p></li>
<li><p><strong>high.throughput.data.type</strong>, one of the five supported high-throughput studies: RNA-seq, MicroRNA-seq, microarray.mRNA, microarray.microRNA and methylation.</p></li>
</ul>
<p>Function also contains two other options:</p>
<ul>
<li><p><strong>cancerCode</strong>, if set as <code>TRUE</code>, function will use the standard abbreviated cancer names instead of complete cancer names. For example, <code>laml_tcga_pub</code> is the shortened name for <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>.</p></li>
<li><p><strong>validateGenes</strong> a logical value that, if set to be ‘TRUE’, function will check each cancer study to find whether or not each gene has a record. If the given cancer doesn’t have a record for specific gene, it checks for alternative gene names that cbioportal might use instead of the given gene name.</p></li>
</ul>
<p>In the following example, <em>genes</em> consists of two gene groups K.demethylases and K.acetyltransferases, <em>submissionName</em> is <code>test</code>, <em>cancername</em> has complete names of five cancer studies and the desired high-throughput study is <code>RNA-seq</code>.</p>
<pre><code>genes &lt;- list(K.demethylases = c(&quot;KDM1A&quot;, &quot;KDM1B&quot;, &quot;KDM2A&quot;), K.acetyltransferases = c(&quot;CLOCK&quot;, &quot;CREBBP&quot;, &quot;ELP3&quot;, &quot;EP300&quot;))


cancernames &lt;- c(&quot;Acute Myeloid Leukemia (TCGA, Provisional)&quot;, &quot;Adrenocortical Carcinoma (TCGA, Provisional)&quot;, &quot;Bladder Urothelial Carcinoma (TCGA, Provisional)&quot;, &quot;Brain Lower Grade Glioma (TCGA, Provisional)&quot;, &quot;Breast Invasive Carcinoma (TCGA, Provisional)&quot;)


obntainMultipleStudies(genes, &quot;test2&quot;, cancernames, &quot;RNA-seq&quot;)</code></pre>
<p>The following code shows how to create a matrix with desired names for <code>cancernames</code>:</p>
<pre><code>cancernames &lt;- matrix(c(&quot;Acute Myeloid Leukemia (TCGA, Provisional)&quot;, &quot;acute myeloid leukemia&quot;, &quot;Adrenocortical Carcinoma (TCGA, Provisional)&quot;, &quot;adrenocortical carcinoma&quot;, &quot;Bladder Urothelial Carcinoma (TCGA, Provisional)&quot;, &quot;bladder urothelial carcinoma&quot;, &quot;Brain Lower Grade Glioma (TCGA, Provisional)&quot;, &quot;brain lower grade glioma&quot;, &quot;Breast Invasive Carcinoma (TCGA, Provisional)&quot;,  &quot;breast invasive carcinoma&quot;), nrow = 5, ncol=2 , byrow = TRUE)</code></pre>
<p>An example of how more options for <code>processMultipleStudies()</code> can be altered:</p>
<pre><code>ObtainMultipleStudies(genes, &quot;test2&quot;, cancernames, &quot;RNA-seq&quot;, cancerCode = TRUE, validateGenes = FALSE)</code></pre>
<p>Function starts by checking whether or not the requested data has already been obtained and, if the requested data exists, it prints the following message and then prevents further processing.</p>
<pre><code>[1] &quot;--- Function 'obtainMultipleStudies()' was skipped: the requested data already exist ---&quot;</code></pre>
<p>If there is a change in function input data, e.g. gene names, or the requested data has not been obtained, it proceeds by getting the data and informs the user with a progress bar.</p>
<pre><code>[1] &quot;*** Obtaining the requested data for test2 ***&quot;
|===================================================================| 100%</code></pre>
<p>The output is stored as a BiocFileCache object, which its name is the combination of <code>bfc_</code> and <em>submissionName</em>. Accordingly, in our example the BiocFileCache name is <code>bfc_test2</code>.</p>
</div>
<div id="automatedstatistics" class="section level2">
<h2>automatedStatistics()</h2>
<p>This function calculates the statistics for the data obtained by <code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code> functions. Based on user’s preference, these statistics can include <em>frequency percentage</em>, <em>frequency ratio</em>, <em>mean value</em> and <em>median value</em> of samples greater than specific value. Furthermore, it can look for the genes that comprise the highest values in each cancer and list the top 5 genes for <em>frequency percentage</em>, <em>mean value</em> and <em>median value</em>.</p>
<p>It requires at least two arguments:</p>
<ul>
<li><p><strong>submissionName</strong>, a character string containing name of interest. It is used for naming the process and should be the same name used for either of <code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code> functions.</p></li>
<li><p><strong>obtainedDataType</strong>, a character string that identifies the type of input data produced by the previous function. Two options are availabe: ‘single study’ for <code>obtainOneStudy()</code> and <code>multiple studies</code> for <code>obtainMultipleStudies()</code>. The function uses <code>obtainedDataType</code> and <code>submissionName</code> to construct the name of the BiocFileCach object and then finds the appropiate data inside it. Default value is <code>multiple studies</code>.</p></li>
</ul>
<p>Function also contains four other options:</p>
<ul>
<li><p><strong>calculate</strong>, a character vector that containes the statistical precedures user desires. Default input is <code>c(&quot;frequencyPercentage&quot;, &quot;frequencyRatio&quot;, &quot;Mean.Value&quot;, &quot;medianValue&quot;)</code>. This will tell the function to compute the followings:</p>
<ul>
<li><p><em>frequencyPercentage</em>, which is the number of samples having the value greather than specific cutoff divided by the total sample size for every study / study subgroup</p></li>
<li><p><em>frequency ratio</em>, which shows the number of selected samples divided by the total number of samples that give the frequency percentage. It shows the selecected and total sample sizes.</p></li>
<li><p><em>Mean Value</em>, which contains mean value of selected samples for each study.</p></li>
<li><p><em>Median Value</em>, which shows the median value of selected samples for every study.</p></li>
</ul></li>
<li><p><strong>topGenes</strong>, a logical value that, if <code>TRUE</code>, causes the function to calculate the five genes comprising the highest values in <em>frequencyPercentage</em>, <em>Mean.Value</em> and <em>medianValue</em> if they are included in <strong>calculate</strong>.</p></li>
<li><p><strong>cutoff</strong>, a number used to limit samples to those that are greather than this number (cutoff). The default value for methylation data is 0.6 while gene expression studies use default value of 2. For methylation studies, it is <em>observed/expected ratio</em>, for the rest, it is <em>z-score</em>. To change the cutoff to any desired number, change the option to <code>cutoff = desiredNumber</code> in which desiredNumber is the number of interest.</p></li>
<li><p><strong>round</strong>, a logical value that, if set to be <code>TRUE</code>, will force the function to round all the calculated values to two decimal places. The default value is <code>TRUE</code>.</p></li>
</ul>
<p>In the following example, <em>submissionName</em> is <code>test</code>, and the <em>obtainedDataType</em> is <code>multiple studies</code>.</p>
<pre><code>automatedStatistics(&quot;test&quot;, obtainedDataType = &quot;multiple studies&quot;)
</code></pre>
<p>The following is the same as previous but excludes <em>mean value</em> and <em>median value</em>. Not that top genes for these two will also be ignored.</p>
<pre><code>automatedStatistics(&quot;test&quot;, obtainedDataType = &quot;multiple studies&quot;, calculate = c(&quot;frequencyPercentage&quot;, &quot;frequencyRatio&quot;))
</code></pre>
<p>Function starts by checking whether or not output data of the previous function exists. If not, an error message will appear:</p>
<pre><code>Error: Please run one of the obtainSingleStudy() or obtainMultipleStudies() functions first</code></pre>
<p>If it can find the desired data but the last execution of the previous function (either <code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code>) was skipped, it then checks its own parameters. If the parameters are the same as the last time <em>automatedStatistics()</em> has been run, it will also skip, printing the following message:</p>
<pre><code>[1] &quot;--- Function 'automatedStatistics()' was skipped: the requested data already exist ---&quot;</code></pre>
<p>If there is a change in function input data, e.g. chenge in cutoff value, or the requested data does not exist, it continues by computing the requested statistics and informs the user with a progress bar.</p>
<pre><code>[1] &quot;*** Performing the requested statistical analyses for test2 ***&quot;
|===================================================================| 100%</code></pre>
<p>The output is stored as a new section of the previous BiocFileCache object which is created by one of the two former functions (<code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code>).</p>
</div>
<div id="heatmapoutput" class="section level2">
<h2>heatmapOutput()</h2>
<p>This function prepares heatmap for <em>frequency percentage</em>, <em>mean value</em> and <em>median value</em> data provided by <code>automatedStatistics()</code> function.</p>
<p>It requires at least one arguments:</p>
<ul>
<li><strong>submissionName</strong>, a character string containing name of interest. It is used for naming the process and should be the same name used for either of <code>obtainOneStudy()</code> or <code>obtainMultipleStudies()</code> functions.</li>
</ul>
<p>Function also contains thirteen other options:</p>
<ul>
<li><p><strong>shortenStudyNames</strong> a logical value, when equals to <code>TRUE</code>, function will removes the end part of cancer names aiming to shorten them. The removed segment usually contains the name of scientific group that has conducted the experiment.</p></li>
<li><p><strong>genelimit</strong> if large number of genes exist in at least one gene group, this option can be used to limit the number of genes that are shown on heatmap. For instance, <code>genelimit=50</code> will limit the heatmap to 50 genes that show the most variation across multiple study / study subgroups. The default value is <code>none</code>.</p></li>
<li><p><strong>resolution</strong> This option can be used to adjust the resolution of the output heatmaps as ‘dot per inch’. The defalut resolution is 600.</p></li>
<li><p><strong>RowCex</strong> a number that specifies letter size in heatmap row names.</p></li>
<li><p><strong>ColCex</strong> a number that specifies letter size in heatmap column names.</p></li>
<li><p><strong>heatmapMargines</strong> a numeric vector that can be used to set heatmap margins. The default value is <code>heatmapMargines=c(15,07)</code>.</p></li>
<li><p><strong>angleForYaxisNames</strong> a number that determines the angle with which the studies/study subgroups names are shown in heatmaps. The angle is based on degree. The default value is 45.</p></li>
<li><p><strong>heatmapColor</strong> a character string that defines heatmap color. The default value is “RdBu”. “redgreen” is also a popular color in genomic studies. To see the rest of colors, please type <code>library(RColorBrewer)</code> and then <code>display.brewer.all()</code>.</p></li>
<li><p><strong>reverseColor</strong> a logical value that reverses the color gradiant for heatmap(s).</p></li>
<li><p><strong>transposedHeatmap</strong> a logical value that changes row and columns of heatmap.</p></li>
<li><p><strong>simplify</strong> a logical value that tells the function whether or not to change values under <em>simplifictionCuttoff</em> to zero. The purpose behind this option is to facilitate seeing candidate genes. Therefore, it is not suited for publications.</p></li>
<li><p><strong>simplifictionCuttoff</strong> a logical value that, if <code>simplify.visulization = TRUE</code>, needs to be set as a desired cuttoff <em>simplify.visulization</em>. It has the same unit as <em>cutoff</em>.</p></li>
<li><p><strong>genesToDrop</strong> a character vector. Gene names within this vector will be removed from heatmap.</p></li>
</ul>
<p>In the following example, <em>submissionName</em> is <code>test</code>.</p>
<pre><code>heatmapOutput(submissionName = &quot;test&quot;)
</code></pre>
<p>The following is the same as previous but uses more options.</p>
<pre><code>heatmapOutput(&quot;test&quot;, shortenStudyNames = TRUE, heatmapMargines = c(13,5), heatmapColor = &quot;redgreen&quot;, genesToDrop = c(&quot;PVT1&quot;, &quot;SNHG6&quot;), RowCex = 1, ColCex = 1, reverseColor = FALSE)
</code></pre>
<p>Function starts by checking whether or not output data of the previous function exists. If not, an error message will appear:</p>
<pre><code>Error: Please run one of the obtainSingleStudy() or obtainMultipleStudies() functions and then the automatedStatistics() function</code></pre>
<p>Otherwise, it continues by preparing and storing the requested heatmaps and informs the user with a progress bar.</p>
<pre><code>[1] &quot;*** Preparing the requested heatmaps for test ***&quot;
|===================================================================| 100%</code></pre>
<p>If there is no change in the options or the requested heatmaps already exist, it avoids storing the heatmaps. The number of skipped heatmaps is then printed:</p>
<pre><code>&quot;--- 1 out of 1 heatmaps was skipped: It already exists. ---&quot;</code></pre>
</div>
<div id="process.multiple.studies" class="section level2">
<h2>process.multiple.studies()</h2>
<p>This function Obtains the requested data for the given genes across the multiple cancer studies. It can check whether or not all genes are included in cancer studies and, if not, looks for the alternative gene names. The main part of function calculates frequency percentage, frequency ratio, mean expression and median of samples greater than specific value. Furthermore, it looks for the genes that comprise the highest values in each cancer. The function has been divided into two main segments. The first segment obtains and processes the desired data while the second segment generates heatmap and excel files.</p>
<p>The function code along with all available options:</p>
<pre><code>process.multiple.studies(genes, cancers, high.throughput.data.type, data.presented.as = c(&quot;Frequency.Percentage&quot;, &quot;Frequency.Ratio&quot;, &quot;Mean.Value&quot;, &quot;Median&quot;), shorteded.cancer.names = TRUE, genelimit=&quot;none&quot;, resolution=600, RowCex=0.8, ColCex=0.8, heatmapMargines=c(24,17), cutoff=2, angle.for.heatmap.cancernames=45, heatmap.color = &quot;RdBu&quot;, reverse.heatmap.color = TRUE, rewrite.output.list = TRUE, round=TRUE, top.genes = TRUE,  validate.genes = TRUE, Use.CancerCode.as.Name = FALSE, simplify.visulization=FALSE, simplifiction.cuttoff=FALSE)</code></pre>
<p>The function requires at least three variables:</p>
<ul>
<li><strong>genes</strong>, a list that contains at least one group of genes. There is no limit for the number of gene groups, users can set as many as gene groups they desire.</li>
<li><strong>cancernames</strong>, a character vector or a matrix possessing names of desired cancer studies. The character vector contains standard cancer names that can be found on cbioportal.org, such as <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>. Alternatively, a matrix can be used if users prefer user-defined cancer names, in which the first column of matrix comprises the standard cancer names while the second column must contain the desired cancer names.</li>
<li><strong>high.throughput.data.type</strong>, one of the five supported high-throughput studies: RNA-seq, MicroRNA-seq, microarray.mRNA, microarray.microRNA and methylation.</li>
</ul>
<p>Function contains other options that can be used for a verity of needs:</p>
<ul>
<li><strong>data.presented.as</strong> tells the function what to compute. The default is <code>c(&quot;Frequency.Percentage&quot;, &quot;Frequency.Ratio&quot;, &quot;Mean.Value&quot;, &quot;Median&quot;)</code> which causes the function to consider all the four possible statistical procedures. Alternatively, users can decrease the number of requested statistical procedures as long as at least one procedure is given to the function.</li>
<li><strong>shorteded.cancer.names</strong> is used to remove the end part of cancer names aiming to shorten them. The removed segment usually contains the name of scientific group that has conducted the experiment.</li>
<li><strong>genelimit</strong> can be employed if large number of genes exists in at least one gene group. This option limits the number of genes to be shown on heatmap. For instance, <code>genelimit=50</code> will limit the heatmap to show 50 genes showing the most variation. The default value is <code>none</code>.</li>
<li><strong>resolution</strong> adjusts the resolution of the output heatmaps as <em>dot per inch</em>. The default value is 600.</li>
<li><strong>RowCex</strong> specifies letter size in heatmap row names.</li>
<li><strong>ColCex</strong> specifies letter size in heatmap column names.</li>
<li><strong>heatmapMargines</strong> sets heatmap margins. The default value is <code>heatmapMargines=c(15,07)</code>.</li>
<li><strong>cutoff</strong> used to limit samples to those that are greater than specific number (cutoff). The default value for methylation data is 0.6 while gene expression studies use default value of 2. For methylation studies, the unit is <code>observed/expected ratio</code>, for the rest, it is <code>z-score</code>- function can understand the correct unit. TO change the cutoff to any desired number, change the option to <code>cutoff = desiredNumber</code> in which <em>desiredNumber</em> is the number of interest.</li>
<li><strong>angle.for.heatmap.cancernames</strong> determines the angle with which the cancer names are shown in heatmaps. The default value is 45 degree.</li>
<li><strong>heatmap.color</strong> can be used to set heatmap colors. The default value is “RdBu”. “redgreen” is also a popular color in genomic studies. To see the rest of colors, please type <code>display.brewer.all()</code>.</li>
<li><strong>reverse.heatmap.color</strong> can reverse the color gradient for heatmap.</li>
<li><strong>rewrite.output.list</strong> This option can be used to modify heatmap, e.g. change margin, without obtaining data from internet again. If set to <code>false</code>, function will use the values previously stored in the global environment (user’s workspace) to draw heatmaps and save an excel file. The default value is <code>TRUE</code>.</li>
<li><strong>round</strong> forces the function to round all the calculated values to two decimal places. The default value is <code>TRUE</code>.</li>
<li><strong>top.genes</strong>, if set as <code>TRUE</code>, cause the function to create three data.frame that contain the gene names with the highest values for each cancer. To get all the three data.frames, <code>Frequency.Percentage</code>, <code>Mean.Value</code> and <code>Median</code> must have been included for <code>data.presented.as</code>.</li>
<li><strong>validate.genes</strong>, if set to be <code>TRUE</code>, function will checks each cancer study to finds whether or not each gene has a record. If a cancer study doesn’t have a record for specific gene, it checks for alternative gene names that cbioportal might use instead of the given gene name.</li>
<li><strong>Use.CancerCode.as.Name</strong>, if set as <code>TRUE</code>, function will use the standard abbreviated cancer names instead of complete cancer names. For example, <code>laml_tcga_pub</code> is the shortened name for <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>.</li>
<li><strong>simplify.visulization</strong>, if <code>TRUE</code>, function alters the values under <code>simplifiction.cuttoff</code> to zero. It only affects heatmaps to assist finding the candidate genes faster. Therefore, it is not suited for publications.</li>
<li><strong>simplifiction.cuttoff</strong> a logical value that, if <code>simplify.visulization = TRUE</code>, needs to be set as a desired cutoff for <code>simplify.visulization</code>. It has the same unit as <code>cutoff</code>.</li>
</ul>
<p>In the following example, <em>genes</em> consists of two gene groups K.demethylases and K.acetyltransferases, <em>cancernames</em> has complete names of five cancer studies and the desired high-throughput study is RNA-seq.</p>
<pre><code>genes &lt;- list(K.demethylases = c(&quot;KDM1A&quot;, &quot;KDM1B&quot;, &quot;KDM2A&quot;), K.acetyltransferases = c(&quot;CLOCK&quot;, &quot;CREBBP&quot;, &quot;ELP3&quot;, &quot;EP300&quot;))


cancernames &lt;- c(&quot;Acute Myeloid Leukemia (TCGA, Provisional)&quot;, &quot;Adrenocortical Carcinoma (TCGA, Provisional)&quot;, &quot;Bladder Urothelial Carcinoma (TCGA, Provisional)&quot;, &quot;Brain Lower Grade Glioma (TCGA, Provisional)&quot;, &quot;Breast Invasive Carcinoma (TCGA, Provisional)&quot;)


process.multiple.studies(genes, cancernames, &quot;RNA-seq&quot;)</code></pre>
<p>The following code shows how to create a matrix with desired names for <code>cancernames</code>:</p>
<pre><code>cancernames &lt;- matrix(c(&quot;Acute Myeloid Leukemia (TCGA, Provisional)&quot;, &quot;acute myeloid leukemia&quot;, &quot;Adrenocortical Carcinoma (TCGA, Provisional)&quot;, &quot;adrenocortical carcinoma&quot;, &quot;Bladder Urothelial Carcinoma (TCGA, Provisional)&quot;, &quot;bladder urothelial carcinoma&quot;, &quot;Brain Lower Grade Glioma (TCGA, Provisional)&quot;, &quot;brain lower grade glioma&quot;, &quot;Breast Invasive Carcinoma (TCGA, Provisional)&quot;,  &quot;breast invasive carcinoma&quot;), nrow = 5, ncol=2 , byrow = TRUE)</code></pre>
<p>An example of how more options for <code>process.multiple.studies()</code> can be altered:</p>
<pre><code>process.multiple.studies(genes, cancernames, &quot;RNA-seq&quot;, data.presented.as = c(&quot;Frequency.Percentage&quot;, &quot;Frequency.Ratio&quot;, &quot;Mean.Value&quot;), shorteded.cancer.names = FALSE, resolution=300, RowCex=1, ColCex=1, heatmapMargines=c(15,5), cutoff=1.5, angle.for.heatmap.cancernames=30, heatmap.color = &quot;redgreen&quot;)</code></pre>
<div id="first-segment-of-function-obtaining-and-processing-the-requested-data" class="section level3">
<h3>First segment of function: obtaining and processing the requested data</h3>
<p>Just like the previous function, <code>process.multiple.studies()</code> starts by printing that it is obtaining and processing the the requested data. Then, a progress bar will appear in the next line.</p>
<pre><code>[1] &quot;[[[ Obtaining and processing the requested data for K.demethylases genes ]]]&quot;
  |                                                                 |   0%          </code></pre>
<p>The progress bar gradually proceeds to 100%. Large number of genes or cancer studies may prolong the processing time. Upon finishing, another message is printed on console, the function creates and stores a list in the R’s global environment with an informative name.</p>
<pre><code>[1] &quot;[[[ Obtaining and processing the requested data for K.demethylases genes ]]]&quot;
  |=================================================================| 100%
[1] &quot;Preparing a 'list' to fill with the requested data&quot;</code></pre>
<p>Based on user’s preference, this list can include some or all of the following groups:</p>
<ul>
<li><strong>Validation.Results</strong>, a character matrix that shows whether or not different cancer studies possess record for various genes. If a cancer study doesn’t have a record for a gene but uses an alternative name, it is shown in the matrix.</li>
<li><strong>Frequency.Percentage</strong>, a numerical matrix that is obtained by dividing the number of samples with values greater than specific cutoff by the total sample size for each cancer study.</li>
<li><strong>Top.Genes.of.Frequency.Percentage</strong>, a data.frame that shows genes with highest value in Frequency.Percentage. It contains name of the gene showing the highest value for every cancer study plus the corresponding value.</li>
<li><strong>Frequency.Ratio</strong>, a character matrix of fractions. Each fraction contains the number of samples with values greater than specific cutoff as a numerator and the total sample size for each cancer study as a denominator.</li>
<li><strong>Mean.Value</strong>, a numerical matrix that shows mean value of samples greater than specific cutoff.</li>
<li><strong>Top.Genes.of.Genes.Expression</strong>, a data.frame that shows genes with highest value in Mean.Value. It contains name of the gene showing the highest value for every cancer study plus the corresponding value.</li>
<li><strong>Median</strong>, a numerical matrix that shows median value of samples greater than specific cutoff.</li>
<li><strong>Top.Genes.of.Median</strong>, a data.frame that shows genes with highest value in Median. It contains name of the gene showing the highest value for every cancer study plus the corresponding value.</li>
</ul>
</div>
<div id="second-segment-of-function-generating-three-heatmaps-and-one-excel-file" class="section level3">
<h3>Second segment of function: generating three heatmaps and one excel file</h3>
<p>The second part of function saves all subgroups of the list, generated by first part of the function, in the different tabs of an excel file. In addition, three heatmaps are generated for Frequency.Percentage, Mean.Value and Median subgroups of the list; based on user’s preference. During the process, a progress bar informs the user about the amount of work that have performed. When finished, a message appears on console that shows the directory where output files can be found.</p>
<pre><code>[1] &quot;Saving the output files&quot;
  |=================================================================| 100%
[1] &quot;Requested excel and PNG files were saved in Z:/cBioAutomatedTools/Source code/cBioAutomatedTools/1. K-demethylases&quot;</code></pre>
<hr />
<p>The two mentioned steps, then, start for <code>K.acetyltransferases</code> gene group, ultimately saving the output files in a separate folder.</p>
<pre><code>[1] &quot;[[[ Obtaining and processing the requested data for K.acetyltransferases genes ]]]&quot;
  |=================================================================| 100%
[1] &quot;Preparing a 'list' to fill with the requested data&quot;
[1] &quot;Saving the output files&quot;
  |=================================================================| 100%
[1] &quot;Requested excel and PNG files were saved in Z:/cBioAutomatedTools/Source code/cBioAutomatedTools/2. K.acetyltransferases&quot;</code></pre>
<p>If users intend to modify the characteristics of generated heatmaps, e.g. change color, function can use the information that is previously stored in the global environment, generating the outputs files without using the internet. To do so, set the <code>rewrite.output.list</code> to FLASE. For instance, the following code can be used to change heatmap color to red-green color gradient:</p>
<pre><code>process.multiple.studies(genes, cancernames, &quot;RNA-seq&quot;, heatmap.color = &quot;redgreen&quot;, rewrite.output.list = FALSE)</code></pre>
</div>
</div>
<div id="process.one.study" class="section level2">
<h2>process.one.study()</h2>
<p>This function Obtains the requested data for the given genes across the multiple subgroups of a cancer study. It is very similar in what it does, to <code>process.multiple.studies()</code> as well as using a two-step procedure.</p>
<p>The function code along with all available options:</p>
<pre><code>process.one.study(genes, cancername, high.throughput.data.type, data.presented.as = c(&quot;Frequency.Percentage&quot;, &quot;Frequency.Ratio&quot;, &quot;Mean.Value&quot;, &quot;Median&quot;), transposedHeatmap=FALSE, desired.case.list=&quot;None&quot;, genelimit=&quot;none&quot;, resolution=600, RowCex=0.8, ColCex=0.8, heatmapMargines=c(10,10), cutoff=&quot;default&quot;, angle.for.heatmap.cancernames=45, heatmap.color = &quot;RdBu&quot;, reverse.heatmap.color = TRUE, rewrite.output.list = TRUE, round=TRUE, top.genes = TRUE, validate.genes = TRUE, simplify.visulization=FALSE, simplifiction.cuttoff=FALSE)</code></pre>
<p>The function requires at least three variables:</p>
<ul>
<li><strong>genes</strong>, a list that contains at least one group of genes. There is no limit for the number of gene groups, users can set as many as gene groups they desire.</li>
<li><strong>cancername</strong>, a character string showing the desired cancer name. It is an standard cancer study name that can be found on cbioportal.org, such as <code>Acute Myeloid Leukemia (TCGA, NEJM 2013)</code>.</li>
<li><strong>high.throughput.data.type</strong>, one of the five supported high-throughput studies: RNA-seq, MicroRNA-seq, microarray.mRNA, microarray.microRNA and methylation.</li>
</ul>
<p>Function contains other options, many of whom are the same as previous function:</p>
<ul>
<li><strong>data.presented.as</strong> tells the function what to compute. The default is <code>c(&quot;Frequency.Percentage&quot;, &quot;Frequency.Ratio&quot;, &quot;Mean.Value&quot;, &quot;Median&quot;)</code> which causes the function to consider all the four possible statistical procedures. Alternatively, users can decrease the number of requested statistical procedures as long as at least one procedure is given to the function.</li>
<li><strong>transposedHeatmap</strong>, if TRUE, function transposes the rows and columns of heatmaps.</li>
<li><strong>desired.case.list</strong> a numeric vector that contains the index of desired cancer subgroups, if user knows index of desired subgroups. If set to be <code>FALSE</code>, function will ask the user to enter them during the process. The default value is <code>none</code>.</li>
<li><strong>genelimit</strong> can be employed if large number of genes exists in at least one gene group. This option limits the number of genes to be shown on heatmap. For instance, <code>genelimit=50</code> will limit the heatmap to show 50 genes showing the most variation. The default value is <code>none</code>.</li>
<li><strong>resolution</strong> adjusts the resolution of the output heatmaps as <em>dot per inch</em>. The default value is 600.</li>
<li><strong>RowCex</strong> specifies letter size in heatmap row names.</li>
<li><strong>ColCex</strong> specifies letter size in heatmap column names.</li>
<li><strong>heatmapMargines</strong> sets heatmap margins. The default value is <code>heatmapMargines=c(15,07)</code>.</li>
<li><strong>cutoff</strong> used to limit samples to those that are greater than specific number (cutoff). The default value for methylation data is 0.6 while gene expression studies use default value of 2. For methylation studies, the unit is <code>observed/expected ratio</code>, for the rest, it is <code>z-score</code>- function can understand the correct unit. TO change the cutoff to any desired number, change the option to <code>cutoff = desiredNumber</code> in which <em>desiredNumber</em> is the number of interest.</li>
<li><strong>angle.for.heatmap.cancernames</strong> determines the angle with which the cancer names are shown in heatmaps. The default value is 45 degree.</li>
<li><strong>heatmap.color</strong> can be used to set heatmap colors. The default value is “RdBu”. “redgreen” is also a popular color in genomic studies. To see the rest of colors, please type <code>display.brewer.all()</code>.</li>
<li><strong>reverse.heatmap.color</strong> can reverse the color gradient for heatmap.</li>
<li><strong>rewrite.output.list</strong> This option can be used to modify heatmap, e.g. change margin, without obtaining data from internet again. If set to <code>false</code>, function will use the values previously stored in the global environment (user’s workspace) to draw heatmaps and save an excel file. The default value is <code>TRUE</code>.</li>
<li><strong>round</strong> forces the function to round all the calculated values to two decimal places. The default value is <code>TRUE</code>.</li>
<li><strong>top.genes</strong>, if set as <code>TRUE</code>, cause the function to create three data.frame that contain the gene names with the highest values for each cancer. To get all the three data.frames, <code>Frequency.Percentage</code>, <code>Mean.Value</code> and <code>Median</code> must have been included for <code>data.presented.as</code>.</li>
<li><strong>validate.genes</strong>, if set to be <code>TRUE</code>, function will checks each cancer study to finds whether or not each gene has a record. If a cancer study doesn’t have a record for specific gene, it checks for alternative gene names that cbioportal might use instead of the given gene name.</li>
<li><strong>simplify.visulization</strong>, if <code>TRUE</code>, function alters the values under <code>simplifiction.cuttoff</code> to zero. It only affects heatmaps to assist finding the candidate genes faster. Therefore, it is not suited for publications.</li>
<li><strong>simplifiction.cuttoff</strong> a logical value that, if <code>simplify.visulization = TRUE</code>, needs to be set as a desired cutoff for <code>simplify.visulization</code>. It has the same unit as <code>cutoff</code>.</li>
</ul>
<p>In the following example, <em>genes</em> consists of two gene groups K.demethylases and K.acetyltransferases, <em>cancername</em> is <code>Breast Invasive Carcinoma (TCGA, Cell 2015)</code>. and the desired high-throughput study is RNA-seq.</p>
<pre><code>genes &lt;- list(K.demethylases = c(&quot;KDM1A&quot;, &quot;KDM1B&quot;, &quot;KDM2A&quot;), K.acetyltransferases = c(&quot;CLOCK&quot;, &quot;CREBBP&quot;, &quot;ELP3&quot;, &quot;EP300&quot;))

process.one.study(genes, &quot;Breast Invasive Carcinoma (TCGA, Cell 2015)&quot;, &quot;RNA-seq&quot;)</code></pre>
<p>Here is example of how more options for <code>process.multiple.studies()</code> can be altered, including index of desired subgroups for <code>desired.case.list</code>:</p>
<pre><code>process.one.study(genes, cancernames, &quot;RNA-seq&quot;, desired.case.list = c(3,4,5), data.presented.as = c(&quot;Frequency.Percentage&quot;, &quot;Frequency.Ratio&quot;, &quot;Mean.Value&quot;), resolution=300, RowCex=1, ColCex=1, heatmapMargines=c(15,5), cutoff=1.5, angle.for.heatmap.cancernames=30, heatmap.color = &quot;redgreen&quot;)</code></pre>
<div id="first-segment-of-function-obtaining-and-processing-the-requested-data-1" class="section level3">
<h3>First segment of function: obtaining and processing the requested data</h3>
<p><code>process.multiple.studies()</code> starts by printing that it is obtaining and processing the the requested data.</p>
<pre><code>[1] &quot;[[[ Obtaining and processing the requested data for K.demethylases genes ]]]&quot;</code></pre>
<p>Then, assuming <code>desired.case.list = &quot;none&quot;</code>, all subgroups of the requested cancer study appear on console, asking the user to chose the index of desired subgroups:</p>
<pre><code>[1] &quot;Please enter the numeric index of desired case list(s) for Breast Invasive Carcinoma (TCGA, Cell 2015) separated by comma. For instance two indexes 1 and 2 should be enterd as 1, 2&quot;
 [1] &quot;1. All Complete Tumors&quot;                                &quot;2. All Tumors&quot;                                        
 [3] &quot;3. ER+ breast tumors&quot;                                  &quot;4. ER- breast tumors&quot;                                 
 [5] &quot;5. Her2-positive breast tumors&quot;                        &quot;6. Invasive Ductal Cancer (Luminal A)&quot;                
 [7] &quot;7. Invasive Ductal Cancer (PAM50 Basal-like)&quot;          &quot;8. Invasive Ductal Cancer (PAM50 Her2-enriched)&quot;      
 [9] &quot;9. Invasive Ductal Cancer (PAM50 Luminal B)&quot;           &quot;10. Invasive Lobular Cancer&quot;                          
[11] &quot;11. Invasive Lobular Cancer (Luminal A)&quot;               &quot;12. Mixed IDC/ILC breast cancer&quot;                      
[13] &quot;13. Other histologies breast cancer&quot;                   &quot;14. Sequenced Tumors&quot;                                 
[15] &quot;15. TCGA Freeze 2015&quot;                                  &quot;16. Triple-negative breast tumors&quot;                    
[17] &quot;17. Tumor Samples with CNA data&quot;                       &quot;18. Tumor Samples with methylation data (HM27)&quot;       
[19] &quot;19. Tumor Samples with methylation data (HM450)&quot;       &quot;20. Tumor Samples with mRNA data (Agilent microarray)&quot;
[21] &quot;21. Tumor Samples with mRNA data (RNA Seq V2)&quot;         &quot;22. Tumor Samples with RPPA data&quot;                     
[23] &quot;23. Tumor Samples with sequencing and CNA data&quot;</code></pre>
<pre><code>Enter the numeric index(es):                  2,3,4,5</code></pre>
<p>Then, a progress bar will appear in the next line. The progress bar gradually proceeds to 100%. Large number of genes may prolong the processing time. Upon finishing, another message is printed on console, the function creates and stores a list in the R’s global environment with an informative name.</p>
<pre><code>  |==================================================================| 100%
[1] &quot;Preparing a 'list' to fill with the requested data&quot;</code></pre>
<p>Based on user’s preference, this list can include some or all of the following groups:</p>
<ul>
<li><strong>Validation.Results</strong>, a character matrix that shows whether or not different cancer studies possess record for various genes. If a cancer study doesn’t have a record for a gene but uses an alternative name, it is shown in the matrix.</li>
<li><strong>Frequency.Percentage</strong>, a numerical matrix that is obtained by dividing the number of samples with values greater than specific cutoff by the total sample size for each cancer study.</li>
<li><strong>Top.Genes.of.Frequency.Percentage</strong>, a data.frame that shows genes with highest value in Frequency.Percentage. It contains name of the gene showing the highest value for every cancer study plus the corresponding value.</li>
<li><strong>Frequency.Ratio</strong>, a character matrix of fractions. Each fraction contains the number of samples with values greater than specific cutoff as a numerator and the total sample size for each cancer study as a denominator.</li>
<li><strong>Mean.Value</strong>, a numerical matrix that shows mean value of samples greater than specific cutoff.</li>
<li><strong>Top.Genes.of.Genes.Expression</strong>, a data.frame that shows genes with highest value in Mean.Value. It contains name of the gene showing the highest value for every cancer study plus the corresponding value.</li>
<li><strong>Median</strong>, a numerical matrix that shows median value of samples greater than specific cutoff.</li>
<li><strong>Top.Genes.of.Median</strong>, a data.frame that shows genes with highest value in Median. It contains name of the gene showing the highest value for every cancer study plus the corresponding value.</li>
</ul>
</div>
<div id="second-segment-of-function-generating-three-heatmaps-and-one-excel-file-1" class="section level3">
<h3>Second segment of function: generating three heatmaps and one excel file</h3>
<p>The second part of function saves all subgroups of the list, generated by first part of the function, in the different tabs of an excel file. In addition, three heatmaps are generated for Frequency.Percentage, Mean.Value and Median subgroups of the list; based on user’s preference. During the process, a progress bar informs the user about the amount of work that have performed. When finished, a message appears on console that shows the directory where output files can be found.</p>
<pre><code>[1] &quot;Saving the output files&quot;
  |==================================================================| 100%
[1] &quot;Requested excel and PNG files were saved in Z:/cBioAutomatedTools/Source code/cBioAutomatedTools/1. K-demethylases&quot;&quot;</code></pre>
<hr />
<p>The two mentioned steps, then, starts for <code>K.acetyltransferases</code> gene group, saving the output files in a separate folder in the end.</p>
<pre><code>[1] &quot;[[[ Obtaining expression data for K.acetyltransferases genes ]]]&quot;
[1] &quot;Please enter the numeric index of desired case list(s) for Breast Invasive Carcinoma (TCGA, Cell 2015) separated by comma. For instance two indexes 1 and 2 should be enterd as 1, 2&quot;
 [1] &quot;1. All Complete Tumors&quot;                                &quot;2. All Tumors&quot;                                        
 [3] &quot;3. ER+ breast tumors&quot;                                  &quot;4. ER- breast tumors&quot;                                 
 [5] &quot;5. Her2-positive breast tumors&quot;                        &quot;6. Invasive Ductal Cancer (Luminal A)&quot;                
 [7] &quot;7. Invasive Ductal Cancer (PAM50 Basal-like)&quot;          &quot;8. Invasive Ductal Cancer (PAM50 Her2-enriched)&quot;      
 [9] &quot;9. Invasive Ductal Cancer (PAM50 Luminal B)&quot;           &quot;10. Invasive Lobular Cancer&quot;                          
[11] &quot;11. Invasive Lobular Cancer (Luminal A)&quot;               &quot;12. Mixed IDC/ILC breast cancer&quot;                      
[13] &quot;13. Other histologies breast cancer&quot;                   &quot;14. Sequenced Tumors&quot;                                 
[15] &quot;15. TCGA Freeze 2015&quot;                                  &quot;16. Triple-negative breast tumors&quot;                    
[17] &quot;17. Tumor Samples with CNA data&quot;                       &quot;18. Tumor Samples with methylation data (HM27)&quot;       
[19] &quot;19. Tumor Samples with methylation data (HM450)&quot;       &quot;20. Tumor Samples with mRNA data (Agilent microarray)&quot;
[21] &quot;21. Tumor Samples with mRNA data (RNA Seq V2)&quot;         &quot;22. Tumor Samples with RPPA data&quot;                     
[23] &quot;23. Tumor Samples with sequencing and CNA data&quot;       
Enter the numeric index(es):                  2,3,4,5
  |==================================================================| 100%
[1] &quot;Preparing a 'list' to fill with the requested data&quot;
[1] &quot;Saving the output files&quot;
  |==================================================================| 100%
[1] &quot;Requested excel and PNG files were saved in Z:/cBioAutomatedTools/Source code/cBioAutomatedTools/2. K-acetyltransferases&quot;</code></pre>
<p>If users intend to modify the characteristics of generated heatmaps, e.g. change color, function can use the information that is previously stored in the global environment, generating the outputs files without using the internet. To do so, set the <code>rewrite.output.list</code> to FLASE. For instance, the following code can be used to change heatmap color to red-green color gradient:</p>
<pre><code>process.one.study(genes, &quot;Breast Invasive Carcinoma (TCGA, Cell 2015)&quot;, &quot;RNA-seq&quot;, heatmap.color = &quot;redgreen&quot;, rewrite.output.list = FALSE)</code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
