#' @title Checking Expression/methylation Profile for various cancer studies.
#'
#' @description This function can prepaire heatmap for 'frequency percentage', 'mean value' and 'median value' based in user's
#' preference.
#'
#' @details
#' \tabular{lllll}{
#' Package: \tab cBioAutomatedTools \cr
#' Type: \tab Package \cr
#' Version: \tab 0.99.0 \cr
#' Date: \tab 2017-06-20 \cr
#' License: \tab Artistic-2.0 \cr
#' }
#'
#' @import gplots RColorBrewer genefilter
#'
#' @usage heatmapOutput(submissionName, desiredTechnique, cutoff = NULL, shortenStudyNames = TRUE, genelimit = "none",
#' resolution = 600, RowCex = 0.8, ColCex = 0.8, heatmapMargines = c(10,10), angleForYaxisNames = 45, heatmapColor = "RdBu",
#' reverseColor = TRUE, transposedHeatmap = FALSE, simplify = FALSE, simplifictionCuttoff = FALSE)
#'
#' @param submissionName a character string containing name of interest. It is used for naming the process.
#'
#' @param desiredTechnique a character string that is one of the following techniques: \code{'RNA-seq'}, \code{'microRNA-Seq'},
#' \code{'microarray.mRNA'}, \code{'microarray.microRNA'} or \code{'methylation'}.
#'
#' @param cutoff a number used to limit samples to those that are greather than specific number (cutoff). The default value for
#' methylation data is 0.6 while gene expression studies use default value of 2. For methylation studies, it is
#' \code{observed/expected ratio}, for the rest, it is \code{z-score}. TO change the cutoff to any desired number, change the
#' option to \code{cutoff = desiredNumber} in which \code{desiredNumber} is the number of interest.
#'
#' @param shortenStudyNames a logical vector. If the value is set as true, function will try to remove the end part of
#' cancer names aiming to shorten them. The removed segment usually contains the name of scientific group that has conducted
#' the experiment.
#'
#' @param genelimit if large number of genes exists in at least one gene group, this option can be use to limit the number of
#' genes to be shown on hitmap. For instance, \code{genelimit=50} will limit the heatmap to 50 genes showing the most variation.
#' The default value is \code{none}.
#'
#' @param resolution a number. This option can be used to adjust the resolution of the output heatmaps as 'dot per inch'.
#' The defalut value is 600.
#'
#' @param RowCex a number that specifies letter size in heatmap row names.
#'
#' @param ColCex a number that specifies letter size in heatmap column names.
#'
#' @param heatmapMargines a numeric vectors that can be use to set heatmap margins. The default value is
#' \code{heatmapMargines=c(15,07)}.
#'
#' @param angleForYaxisNames a number that determines the angle with which the studies/study subgroups names are shown in heatmaps.
#' The default value is 45 degree.
#'
#' @param heatmapColor a character string matches standard color names. The default value is "RdBu". "redgreen" is also a popular
#' color in genomic studies. To see the rest of colors, please type \code{display.brewer.all()}.
#'
#' @param reverseColor a logical value that reverses the color gradiant for heatmap.
#'
#' @param simplify a logical value that tells the function whether or not to change values under
#' \code{simplifiction.cuttoff} to zero. It only affects heatmaps to assist finding the candidate genes faster. Therefore, it is
#' not suited for publications.
#'
#' @param simplifictionCuttoff a logical value that, if \code{simplify.visulization = TRUE}, needs to be set as a desired cuttoff
#' for \code{simplify.visulization}. It has the same unit as \code{cutoff}.
#'
#' @return Based on preference, three heatmaps for \code{Frequency.Percentage}, \code{Mean.Value} and \code{Median.value} can be
#' generated. If more than one group of genes is entered, output for each group will be strored in a separate sub-directory.
#'
#' @examples
#'
#' # Prepairing heatmap for preexist data that is generated by automatedStatistics()
#' heatmapOutput(submissionName = "test", desiredTechnique = "RNA-seq")
#'
#' @author Arman Shahrisa, \email{shahrisa.arman@hotmail.com} [maintainer, copyright holder]
#' @author Maryam Tahmasebi Birgani, \email{tahmasebi-ma@ajums.ac.ir}
#'
#' @export



#########################################################################
#########################################################################
############# Obtain the requested data for multiple Cancer #############
#########################################################################
#########################################################################

xlsxOutput <- function(submissionName){

  ##########################################################################
  ########## Prerequisites

  # Check submissionName

  if(!is.character(submissionName)){

    stop("'submissionName' must be entered as a character string for naming the process")

  }





  ##########################################################################
  ########## Decide whether functions should stops now!

  # Store the new parameteres

  newParameters <-list()

  newParameters$submissionName <- submissionName





  # Check wheather the requested data exists

  if(!exists(paste("Pa.PrData.", submissionName, sep = ""))){

    Stop("Please run automatedStatistics() function first")

  } else{

    oldParam <- get(paste("Pa.PrData.", submissionName, sep = ""))

    desiredTechnique <- oldParam$desiredTechnique

    cutoff <- oldParam$cutoff

    obtainedDataType <- oldParam$obtainedDataType

  }



  # setting the value for cutoff

  if(desiredTechnique == "methylation"){

    cutoff.phrase <- "obs/exp cutoff"

  } else{

    cutoff.phrase <- "z-score cutoff"

  }



  if(exists(paste("Pa.Excel.", submissionName, sep = ""))){

    if(oldParam$HaultOrder == TRUE){

      if(identical(get(paste("Pa.Excel.", submissionName, sep = "")), newParameters)){

        continue <- FALSE

      } else{

        continue <- TRUE

      }

    } else{

      continue <- TRUE

    }

  } else{

    continue <- TRUE

  }





  # Get the previous function's result

  whole.data <- get(paste("PrData.", submissionName, sep = ""))

  # Check gene validation dat

  if(obtainedDataType == "multiple studies"){

    validationType <- "Va.Mu."

  } else if(obtainedDataType == "single study"){

    validationType <- "Va.Si."

  }

  if(exists(paste(validationType, submissionName, sep = ""))){

    validation.data <- get(paste(validationType, submissionName, sep = ""))

  }





  # get the working directory

  parent.directory <- getwd()



  ##########################################################################
  ########## Set the function ready to work

  # Report

  print(paste("***", "Preparing the requested excel file(s) for", submissionName, "***", sep = " "))



  # Count number of skipped excel files

  skipped <- 0





  # Create progressbar

  total.number <- length(whole.data)

  heatmapOutputProgressBar <- txtProgressBar(min = 0, max = total.number, style = 3)

  ExtH <- 0





  ##########################################################################
  ########## Core segment

  # Save heatmaps in separate folder

  for(gr in 1:length(whole.data)){

    # Subset data that can be presented as heatmap

    subset.name <- names(whole.data)[gr]

    subset.data <- whole.data[[gr]]

    sheet.names <- names(subset.data)



    # merge with validation data

    if(exists("validation.data")){

      v.subset.name <- names(validation.data)[gr]

      v.subset.data <- validation.data[[gr]]



      if(identical(subset.name, v.subset.name)){

        subset.data$Validation.Result <- v.subset.data

        subset.data <- subset.data[match(c("Validation.Result", sheet.names), names(subset.data))]

      }

    }







    # Create a directory and set it as desired folder

    child.directory <- paste(gr, ". ", sub(x = subset.name, pattern = "\\.", replacement = "-"), sep="")

    dir.create(paste(parent.directory, child.directory, sep = "/"), showWarnings = FALSE)

    setwd(paste(parent.directory, child.directory, sep = "/"))



    # determine ourput file name

    name.of.excel.file <- paste(gsub(x = subset.name, pattern = "\\.", replacement = "-"), " (",cutoff.phrase, "=", cutoff, ")" , ".xlsx", sep = "")

    # Check if excel file already exists

    if(continue == TRUE & file.exists(name.of.excel.file)){

      file.remove(name.of.excel.file)

      hault <- FALSE

    }else if(continue == TRUE | continue == FALSE & !file.exists(name.of.excel.file)){

      hault <- FALSE

    }else{

      skipped <- skipped + 1

      hault <- TRUE

    }




    for(possible in 1:length(subset.data)){

      if(hault == FALSE){

        # subset statistics

        statistics.data <- subset.data[[possible]]

        name.statistics.data <- names(subset.data)[possible]



        if(name.statistics.data %in% c("Frequency.Percentage", "Mean.Value", "Median.Value")){

          # Replace 'NaN' values with character string

          statistics.data[is.nan(statistics.data)] <- "NaN"

          # Replace 'NA' values with character string

          statistics.data[is.na(statistics.data)] <- "NA"

        }



        if(name.statistics.data %in% c("Top.Genes.of.Frequency.Percentage", "Top.Genes.of.Mean.Value", "Top.Genes.of.Median.Value")){

          # Removing index as rownames

          statistics.data <- statistics.data[, 2:length(ncol(statistics.data))]

        }



        # Saving the expression profile

        write.xlsx(statistics.data, file = name.of.excel.file, sheetName = gsub(x = name.statistics.data, pattern = "\\.", replacement = " "), append = TRUE)

      }

      # Update progressbar

      ExtH <- ExtH + 1

      setTxtProgressBar(heatmapOutputProgressBar, ExtH)

    }

  }

  # Close progressbar

  close(heatmapOutputProgressBar)

  # report number of skipped heatmaps

  if(skipped > 0 & skipped != 1){

    print(paste("--- ", as.character(skipped), " out of ", as.character(total.number)," heatmaps were skipped: They already exist. ---", sep = ""))

  } else if(skipped > 0 & skipped == 1){

    print(paste("--- ", as.character(skipped), " out of ", as.character(total.number)," heatmaps was skipped: It already exist. ---", sep = ""))

  }

  # Store the parameters for this run

  assign(paste("Pa.Excel.", submissionName, sep = ""), newParameters, envir = globalenv())

  # change directory to parent directory

  setwd(parent.directory)

}
